{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Modelado de los residuos restantes de los cristales\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Obtener la secuencia de UniProt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Librerías\n",
    "from modeller import * # licencia modeller necesaria\n",
    "from prody import *\n",
    "import csv\n",
    "from modeller import * # licencia modeller necesaria\n",
    "from modeller.automodel import *\n",
    "from prody import * \n",
    "from Bio import pairwise2\n",
    "import os\n",
    "import glob"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# Secuencia de la CDK2 de UniProt\n",
    "cdk2_P24941 = \"MENFQKVEKIGEGTYGVVYKARNKLTGEVVALKKIRLDTETEGVPSTAIREISLLKELNHPNIVKLLDVIHTENKLYLVFEFLHQDLKKFMDASALTGIPLPLIKSYLFQLLQGLAFCHSHRVLHRDLKPQNLLINTEGAIKLADFGLARAFGVPVRTYTHEVVTLWYRAPEILLGCKYYSTAVDIWSLGCIFAEMVTRRALFPGDSEIDQLFRIFRTLGTPDEVVWPGVTSMPDYKPSFPKWARQDFSKVVPPLDEDGRSLLSQMLHYDPNKRISAKAALAHPFFQDVTKPVPHLRL\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Función find gaps"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Versión marzo 2018 *Anotar las modificaciones en esta versión y comentar lo pasado\n",
    "def find_gaps(seq, r=1): # r es el número de residuos móviles al lado de la ventana del gap\n",
    "    import re\n",
    "    '''Encontrar el número e inicio y final de los gaps en el alineamiento'''\n",
    "    seq_len = len(seq)\n",
    "    gaps = list(re.finditer('[-]+', seq)) # todas las subsecuencias que tienen uno o más guiones\n",
    "    num_gaps = len(gaps); gap_lengths = []\n",
    "    gap_list = []; gap_window = []\n",
    "    # Obtener el índice de inicio y final del gap\n",
    "    for i , gap in enumerate(gaps, 1):\n",
    "        start = gap.start() + 1\n",
    "        end = gap.end()\n",
    "        gap_lengths.append(end - start + 1)\n",
    "        gap_list.append([start, end])\n",
    "        end_right = end if end >= seq_len else end  + r\n",
    "        start_right = start if start == 1 else start - r\n",
    "        gap_window.append([start_right, end_right])\n",
    "    return({\"num_gaps\": num_gaps, \"gap_lengths\":gap_lengths,\n",
    "            \"gap_list\": gap_list, \"gap_window\": gap_window})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'num_gaps': 3,\n",
       " 'gap_lengths': [3, 1, 2],\n",
       " 'gap_list': [[5, 7], [15, 15], [17, 18]],\n",
       " 'gap_window': [[5, 7], [15, 15], [17, 18]]}"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# EJAMPLO:\n",
    "find_gaps(\"LLLL---LLLÑÑÑR-K--L\", r=0)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Obtener la secuencia del PDB del cristal"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "391"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Los IDs de los cristales están en el archivo\n",
    "with open('./CDK2_pdb_391_IDs_P24941.csv', 'r') as f:\n",
    "    reader = csv.reader(f)\n",
    "    cdk2_pdb_ids = list(reader)[0] # Lista de IDs\n",
    "len(cdk2_pdb_ids)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Se ejecuta Modeller para refinamiento por loops"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "@> 2391 atoms and 1 coordinate set(s) were parsed in 0.02s.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Modelando proteína 2wip\n",
      "1\n",
      "self.residue_range('9', '17')\n",
      "\n",
      "check_ali___> Checking the sequence-structure alignment. \n",
      "\n",
      "Implied intrachain target CA(i)-CA(i+1) distances longer than  8.0 angstroms:\n",
      "\n",
      "ALN_POS  TMPL  RID1  RID2  NAM1  NAM2     DIST\n",
      "----------------------------------------------\n",
      "END OF TABLE\n",
      "read_to_681_> topology.submodel read from topology file:        3\n",
      "mdtrsr__446W> A potential that relies on one protein is used, yet you have at\n",
      "              least one known structure available. MDT, not library, potential is used.\n",
      "0 atoms in HETATM/BLK residues constrained\n",
      "to protein atoms within 2.30 angstroms\n",
      "and protein CA atoms within 10.00 angstroms\n",
      "0 atoms in residues without defined topology\n",
      "constrained to be rigid bodies\n",
      "condens_443_> Restraints marked for deletion were removed.\n",
      "              Total number of restraints before, now:    31997    29808\n",
      "64 (of 2398 total) atoms selected for optimization\n",
      "iupac_m_397W> Atoms were not swapped because of the uncertainty of how to handle the H atom.\n",
      "iupac_m_397W> Atoms were not swapped because of the uncertainty of how to handle the H atom.\n",
      "iupac_m_397W> Atoms were not swapped because of the uncertainty of how to handle the H atom.\n",
      "iupac_m_397W> Atoms were not swapped because of the uncertainty of how to handle the H atom.\n",
      "iupac_m_397W> Atoms were not swapped because of the uncertainty of how to handle the H atom.\n",
      "\n",
      "\n",
      ">> ENERGY; Differences between the model's features and restraints:\n",
      "Number of all residues in MODEL                   :      298\n",
      "Number of all, selected real atoms                :     2398      64\n",
      "Number of all, selected pseudo atoms              :        0       0\n",
      "Number of all static, selected restraints         :    29808     593\n",
      "COVALENT_CYS                                      :        F\n",
      "NONBONDED_SEL_ATOMS                               :        1\n",
      "Number of non-bonded pairs (excluding 1-2,1-3,1-4):      167\n",
      "Dynamic pairs routine                             : 5, NATM_SEL x NATM double loop\n",
      "Atomic shift for contacts update (UPDATE_DYNAMIC) :    0.390\n",
      "LENNARD_JONES_SWITCH                              :    6.500   7.500\n",
      "COULOMB_JONES_SWITCH                              :    6.500   7.500\n",
      "RESIDUE_SPAN_RANGE                                :        0   99999\n",
      "NLOGN_USE                                         :       15\n",
      "CONTACT_SHELL                                     :    4.000\n",
      "DYNAMIC_PAIRS,_SPHERE,_COULOMB,_LENNARD,_MODELLER :        T       T       F       F       F\n",
      "SPHERE_STDV                                       :    0.050\n",
      "RADII_FACTOR                                      :    0.820\n",
      "Current energy                                    :         134.4137\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Summary of the restraint violations: \n",
      "\n",
      "   NUM     ... number of restraints.\n",
      "   NUMVI   ... number of restraints with RVIOL > VIOL_REPORT_CUT[i].\n",
      "   RVIOL   ... relative difference from the best value.\n",
      "   NUMVP   ... number of restraints with -Ln(pdf) > VIOL_REPORT_CUT2[i].\n",
      "   RMS_1   ... RMS(feature, minimally_violated_basis_restraint, NUMB).\n",
      "   RMS_2   ... RMS(feature, best_value, NUMB).\n",
      "   MOL.PDF ... scaled contribution to -Ln(Molecular pdf).\n",
      "\n",
      " #                     RESTRAINT_GROUP      NUM   NUMVI  NUMVP   RMS_1   RMS_2         MOL.PDF     S_i\n",
      "------------------------------------------------------------------------------------------------------\n",
      " 1 Bond length potential              :      66       0      0   0.010   0.010      2.0450       1.000\n",
      " 2 Bond angle potential               :      90       0      0   2.303   2.303      9.7067       1.000\n",
      " 3 Stereochemical cosine torsion poten:      38       0      1  47.481  47.481      13.356       1.000\n",
      " 4 Stereochemical improper torsion pot:      28       0      0   1.197   1.197     0.94103       1.000\n",
      " 5 Soft-sphere overlap restraints     :     167       0      0   0.009   0.009      1.6233       1.000\n",
      " 6 Lennard-Jones 6-12 potential       :       0       0      0   0.000   0.000      0.0000       1.000\n",
      " 7 Coulomb point-point electrostatic p:       0       0      0   0.000   0.000      0.0000       1.000\n",
      " 8 H-bonding potential                :       0       0      0   0.000   0.000      0.0000       1.000\n",
      " 9 Distance restraints 1 (CA-CA)      :     117       0      0   0.386   0.386      16.201       1.000\n",
      "10 Distance restraints 2 (N-O)        :     101       3      9   1.079   1.079      53.618       1.000\n",
      "11 Mainchain Phi dihedral restraints  :       0       0      0   0.000   0.000      0.0000       1.000\n",
      "12 Mainchain Psi dihedral restraints  :       0       0      0   0.000   0.000      0.0000       1.000\n",
      "13 Mainchain Omega dihedral restraints:      10       1      1   8.379   8.379      5.7246       1.000\n",
      "14 Sidechain Chi_1 dihedral restraints:       6       0      0  49.039  49.039     0.85933       1.000\n",
      "15 Sidechain Chi_2 dihedral restraints:       4       0      0  95.800  95.800     0.44277       1.000\n",
      "16 Sidechain Chi_3 dihedral restraints:       2       0      0 138.458 138.458      2.2674       1.000\n",
      "17 Sidechain Chi_4 dihedral restraints:       1       0      0   0.178   0.178     0.21763       1.000\n",
      "18 Disulfide distance restraints      :       0       0      0   0.000   0.000      0.0000       1.000\n",
      "19 Disulfide angle restraints         :       0       0      0   0.000   0.000      0.0000       1.000\n",
      "20 Disulfide dihedral angle restraints:       0       0      0   0.000   0.000      0.0000       1.000\n",
      "21 Lower bound distance restraints    :       0       0      0   0.000   0.000      0.0000       1.000\n",
      "22 Upper bound distance restraints    :       0       0      0   0.000   0.000      0.0000       1.000\n",
      "23 Distance restraints 3 (SDCH-MNCH)  :      81       0      0   0.800   0.800      1.6255       1.000\n",
      "24 Sidechain Chi_5 dihedral restraints:       0       0      0   0.000   0.000      0.0000       1.000\n",
      "25 Phi/Psi pair of dihedral restraints:      11       5      1  39.724 128.765      16.111       1.000\n",
      "26 Distance restraints 4 (SDCH-SDCH)  :      38       0      2   2.139   2.139      9.6740       1.000\n",
      "27 Distance restraints 5 (X-Y)        :       0       0      0   0.000   0.000      0.0000       1.000\n",
      "28 NMR distance restraints 6 (X-Y)    :       0       0      0   0.000   0.000      0.0000       1.000\n",
      "29 NMR distance restraints 7 (X-Y)    :       0       0      0   0.000   0.000      0.0000       1.000\n",
      "30 Minimal distance restraints        :       0       0      0   0.000   0.000      0.0000       1.000\n",
      "31 Non-bonded restraints              :       0       0      0   0.000   0.000      0.0000       1.000\n",
      "32 Atomic accessibility restraints    :       0       0      0   0.000   0.000      0.0000       1.000\n",
      "33 Atomic density restraints          :       0       0      0   0.000   0.000      0.0000       1.000\n",
      "34 Absolute position restraints       :       0       0      0   0.000   0.000      0.0000       1.000\n",
      "35 Dihedral angle difference restraint:       0       0      0   0.000   0.000      0.0000       1.000\n",
      "36 GBSA implicit solvent potential    :       0       0      0   0.000   0.000      0.0000       1.000\n",
      "37 EM density fitting potential       :       0       0      0   0.000   0.000      0.0000       1.000\n",
      "38 SAXS restraints                    :       0       0      0   0.000   0.000      0.0000       1.000\n",
      "39 Symmetry restraints                :       0       0      0   0.000   0.000      0.0000       1.000\n",
      "\n",
      "\n",
      "\n",
      "# Heavy relative violation of each residue is written to: 2wip_full.V99990001\n",
      "# The profile is NOT normalized by the number of restraints.\n",
      "# The profiles are smoothed over a window of residues:    1\n",
      "# The sum of all numbers in the file:    1767.8643\n",
      "\n",
      "\n",
      "\n",
      "List of the violated restraints:\n",
      "   A restraint is violated when the relative difference\n",
      "   from the best value (RVIOL) is larger than CUTOFF.\n",
      "\n",
      "   ICSR   ... index of a restraint in the current set.\n",
      "   RESNO  ... residue numbers of the first two atoms.\n",
      "   ATM    ... IUPAC atom names of the first two atoms.\n",
      "   FEAT   ... the value of the feature in the model.\n",
      "   restr  ... the mean of the basis restraint with the smallest\n",
      "              difference from the model (local minimum).\n",
      "   viol   ... difference from the local minimum.\n",
      "   rviol  ... relative difference from the local minimum.\n",
      "   RESTR  ... the best value (global minimum).\n",
      "   VIOL   ... difference from the best value.\n",
      "   RVIOL  ... relative difference from the best value.\n",
      "\n",
      "\n",
      "-------------------------------------------------------------------------------------------------\n",
      "\n",
      "Feature 10                           : Distance restraints 2 (N-O)             \n",
      "List of the RVIOL violations larger than   :       4.5000\n",
      "\n",
      "    #   ICSR  RESNO1/2 ATM1/2   INDATM1/2    FEAT   restr    viol   rviol   RESTR    VIOL   RVIOL\n",
      "    1  15663  17V  34K N   O     128  268    5.77    3.22    2.55    5.04    3.22    2.55    5.04\n",
      "    2  15664  17V  35I N   O     128  276   10.63    7.89    2.73    4.94    7.89    2.73    4.94\n",
      "    3  15665  17V  36R N   O     128  287   11.58    8.84    2.74    4.76    8.84    2.74    4.76\n",
      "\n",
      "-------------------------------------------------------------------------------------------------\n",
      "\n",
      "Feature 25                           : Phi/Psi pair of dihedral restraints     \n",
      "List of the RVIOL violations larger than   :       6.5000\n",
      "\n",
      "    #   ICSR  RESNO1/2 ATM1/2   INDATM1/2    FEAT   restr    viol   rviol   RESTR    VIOL   RVIOL\n",
      "    1   8463  10I  11G C   N      86   88 -149.48 -167.20   19.33    0.57   82.20 -143.95   15.96\n",
      "    1         11G  11G N   CA     88   89 -177.67  174.60                    8.50\n",
      "    2   8464  11G  12E C   N      90   92 -130.21 -117.80   16.68    0.52  -63.60 -175.77   28.32\n",
      "    2         12E  12E N   CA     92   93  147.94  136.80                  -40.30\n",
      "    3   8465  12E  13G C   N      99  101   92.70   78.70   14.57    0.80   82.20  170.86    7.76\n",
      "    3         13G  13G N   CA    101  102 -162.03 -166.10                    8.50\n",
      "    4   8467  14T  15Y C   N     110  112  -77.47  -98.40   24.40    1.47  -63.50  176.22   28.68\n",
      "    4         15Y  15Y N   CA    112  113  140.93  128.40                  -43.40\n",
      "    5   8468  15Y  16G C   N     122  124   79.21   82.20   23.48    0.99  -62.40  159.31   30.13\n",
      "    5         16G  16G N   CA    124  125   31.79    8.50                  -41.20\n",
      "\n",
      "\n",
      "report______> Distribution of short non-bonded contacts:\n",
      "\n",
      "  serious non-bonded atom clash:  1797 1849   2.272\n",
      "\n",
      "DISTANCE1:  0.00 2.10 2.20 2.30 2.40 2.50 2.60 2.70 2.80 2.90 3.00 3.10 3.20 3.30 3.40\n",
      "DISTANCE2:  2.10 2.20 2.30 2.40 2.50 2.60 2.70 2.80 2.90 3.00 3.10 3.20 3.30 3.40 3.50\n",
      "FREQUENCY:     0    0    1    0    1    7   12   65  107  143  195  249  308  340  409\n",
      "\n",
      "\n",
      "<< end of ENERGY.\n",
      "\n",
      ">> Summary of successfully produced models:\n",
      "Filename                          molpdf\n",
      "----------------------------------------\n",
      "2wip_full.B99990001.pdb        134.41365\n",
      "\n",
      "15.625251293182373\n"
     ]
    }
   ],
   "source": [
    "import os\n",
    "import time\n",
    "start = time.time()\n",
    "\n",
    "# Directiros\n",
    "# Mismo para todos\n",
    "struct_dir = './PDB_CDK2_chains/'\n",
    "tail_pdb = '_A.pdb' # extensión del archivo pdb del cristal\n",
    "tail_model = '_full' # epiteto del archivo del modelo final, la extensión .pdb la agrega modeller\n",
    "pdbs_model_dir = \"./Modelos_Full_CDK2_PDB/\" # Carpeta de salida para los modelos\n",
    "num_res_window = 2\n",
    "\n",
    "# *****************************************************************************\n",
    "# Para corregir los modelos que salieron mal en la primera ejecución de modeller\n",
    "# # Conformaciones que necesitan atención:\n",
    "# /home/joel/Documentos/Doctorado/DOCTO_TESIS/Proteinas_Modelo/CDK2/CONFORMACIONES/Prueba_Modeller/Modelos_malos_primera iteración\n",
    "# URGENTE: 1h00, 1oit, 1p2a, 2r3j, 2r3k, 4acm, 4ek6, 4fkj, 5ano, 6ath\n",
    "# LOOPS muy extendidos: 1w98, 5l2w, 5uq3\n",
    "# PRIMERA RONDA:\n",
    "# Se intenta ampliando la ventana de 1 a 2 residuos\n",
    "#cdk2_pdb_ids = ['1h00', '1oit', '1p2a', '2r3j', '2r3k', '4acm', '4ek6', '4fkj', '5ano', '6ath',\n",
    "    #           '1w98', '5l2w', '5uq3']\n",
    "# Originales: 1h00 1oit 1p2a 2r3j 2r3k 4acm 4ek6 4fkj 5ano 6ath 1w98 5l2w 5uq3\n",
    "\n",
    "# SEGUNDA RONDA: (Los que no quedaron en la primera)\n",
    "# 2r3k* 4acm* 4ek6* 6ath* 5uq3*\n",
    "# Más exhaustividad a 1000, residuos a 0\n",
    "\n",
    "# TERCERA RONDA: (Los que no quedaron en la segunda)\n",
    "# 4ek6*: loops entrelazados \n",
    "# 6ath*: N terminal desplegada\n",
    "# 5uq3*: dominio pequeño (N-terminal) sin estructura (critico)\n",
    "# COCLUSIÓN: Descratar los tres debido a que les faltan hasta 16 residuos en la región N-terminal\n",
    "\n",
    "# CUARTA RONDA: (Los que no funcionaban con AD4)\n",
    "# 3s2p, 1oit, 1v1k\n",
    "# window = 2, a.max_var_iterations = 2000 #*500, a.md_level = refine.slow # Nivel del rifinamiento,\n",
    "# a.repeat_optimization = 3 ##Originalmente comentado\n",
    "\n",
    "# QUINTA RONDA: (Los que cuyos loops solapaban con el ligando: altas energias en AD4)\n",
    "# 4bcq, 3pxf, 2wip, 1oiq\n",
    "# window = 1\n",
    "\n",
    "cdk2_pdb_ids = ['2wip'] # No funcionó el refinamiento por loops\n",
    "\n",
    "for pdb_code in cdk2_pdb_ids:\n",
    "    #INPUTS\n",
    "    # variable\n",
    "    code = pdb_code\n",
    "    \n",
    "    ###########################\n",
    "    if os.path.isfile(pdbs_model_dir + code + tail_model + \".pdb\"):\n",
    "        print(\"El modelo ya existe en el directorio\")\n",
    "        continue\n",
    "\n",
    "    ###########################\n",
    "\n",
    "    # Secuencia de la estructura cristalográfica\n",
    "    try:\n",
    "        stc_prot = parsePDB(struct_dir + code + tail_pdb).select('not nonstdaa and resid > 0') # Archivo pdb\n",
    "                                                            # La selección omite residuos negativos\n",
    "            # Además omite los residuos no estandar, con modificaciones postTrad como el TPO (Tirosina fosfatada)\n",
    "        chaid = stc_prot.getChids()[0] # Obtiene el id de la cadena\n",
    "        ref_hv = stc_prot.getHierView()[chaid] # Se obtiene sólo la cadena\n",
    "        seq_cry = ref_hv.getSequence() # Se obtiene la secuencia\n",
    "        \n",
    "        #print(seq_cry)\n",
    "        #print(cdk2_P24941)\n",
    "       \n",
    "    except:\n",
    "        print(\"Error al abrir y modelar:\", pdb_code)\n",
    "        continue\n",
    "\n",
    "    ########################### CONTROL\n",
    "    # Pregunta si la longitud y la secuencias de Uniprot y de la estructura son iguales\n",
    "    same_seq = len(seq_cry) == len(cdk2_P24941) and seq_cry == cdk2_P24941\n",
    "    # si same_seq es verdadero, se omite hacer el modelado pues la estructura está completa\n",
    "    # y se pasa a la siguiente secuencia\n",
    "    \n",
    "    if same_seq:\n",
    "        print(\"La proteína \" + code + \" ya está completa\")\n",
    "        # continue\n",
    "        \n",
    "    print(\"Modelando proteína \" + code)\n",
    "\n",
    "    ###########################\n",
    "    # Obtener la secuecnia de la estructura a modelar y guardarla en un archivo\n",
    "\n",
    "    # Alineamiento global, se penalizan con -10 los gaps abiertos, se obtiene el mejor\n",
    "    alignment = pairwise2.align.globalxs(seq_cry, cdk2_P24941, \n",
    "                                        -10, -0.5, gap_char='-')[0]\n",
    "    # Secuencias del a\n",
    "    algn1_struc = alignment[0] # Secuencia alineada de la estructura cristalográfica\n",
    "    algn2_seq = cdk2_P24941 # alignment[1] == Secuencia completa de UniProt\n",
    "    \n",
    "    # print(algn1_struc)\n",
    "    # print(algn2_seq)\n",
    "\n",
    "    # Nombre de los cabezales de las secuencia\n",
    "    # deben coincidir con el de los archivos de entrada y salida\n",
    "    crys_file_name = code + tail_pdb\n",
    "    model_file_name = code + tail_model\n",
    "\n",
    "    # Necesario: There should be 10 fields separated by colons \":\"\n",
    "    # Please check the file to make sure your sequences end with the '*' character.\n",
    "    # Nomenclaturas de los campos del header: https://salilab.org/modeller/8v2/manual/node176.html\n",
    "\n",
    "    # Headers (la secuencia del cristal va primero, luego la secuencia completa)\n",
    "    struc_header = \"structureX:\" + crys_file_name + \":.:\" + chaid + \":.:\" + chaid + \":.:.:.:\"\n",
    "    seq_header = \"sequence:\" + model_file_name + \":.:.:.:.:.:.:.:\"\n",
    "\n",
    "    # Crea el archivo de alineamiento con la estructura requerida por modeller\n",
    "    alg_filename = code + \".alg\"\n",
    "    with open(alg_filename, \"w\") as handle:\n",
    "        handle.write(\"\\n>P1;%s\\n%s\\n%s*\\n>P1;%s\\n%s\\n%s*\\n\" % (crys_file_name, struc_header, algn1_struc, \n",
    "                                                               model_file_name, seq_header, algn2_seq))\n",
    "\n",
    "    ##########################\n",
    "    gaps = find_gaps(algn1_struc, r = num_res_window)\n",
    "    num_gaps = gaps[\"num_gaps\"]\n",
    "    print(num_gaps)\n",
    "    \n",
    "    ##### Comprueba si hay gaps\n",
    "    if num_gaps != 0:\n",
    "     #   print(\"Se omite hacer el modelado de \" + code + \", la proteína ya está completa o tiene más residuos que la secuencia\")\n",
    "     #   print(\"Longitud de secuencia: \" + str(len(algn2_seq)))\n",
    "     #   print(\"Longitud de la estructura: \" + str(len(algn1_struc)))\n",
    "     #   continue\n",
    "\n",
    "        gap_i = gaps[\"gap_window\"]\n",
    "        s = \"self.residue_range('\" + str(gap_i[0][0]) + \"', '\" + str(gap_i[0][1]) + \"')\"\n",
    "\n",
    "        # A pesar de que sólo haya una, se ejecuta el ciclo, si hay más de una, se crean para las demás\n",
    "        for i in range(1, num_gaps):\n",
    "            s = s + \", \" + \"self.residue_range('\" + str(gap_i[i][0]) + \"', '\" + str(gap_i[i][1]) + \"')\" #nótese la coma\n",
    "        print(s)\n",
    "        ##########################\n",
    "        env = environ()\n",
    "        env.io.atom_files_directory = ['.', struct_dir]\n",
    "\n",
    "        # Modifico loopmodel para fijar los residuos ya existentes en el cristal\n",
    "        # NECESARIO para poder capturar los valores obtenidos por el rango de gaps\n",
    "        # Al estar dentro del loop la identación extra es importante\n",
    "        MyModel_code = \"\"\"\n",
    "class MyModel(automodel):\n",
    "    def select_atoms(self):\n",
    "        return selection(\"\"\" + s + \"\"\")\n",
    "    \"\"\"\n",
    "\n",
    "        exec(MyModel_code) # Se modifica la clase MyModel para agregar la región fijada\n",
    "\n",
    "    a = MyModel(env, alnfile = alg_filename, # Lee el archivo fasta creado y guardado en el directorio actual\n",
    "                      knowns = crys_file_name, # Archivo pdb crys, que coincide con el id en el archivo fasta\n",
    "                      sequence = model_file_name) # Nombre del modelo\n",
    "    a.starting_model= 1\n",
    "    a.ending_model  = 1\n",
    "    #a.library_schedule = autosched.slow ##Originalmente comentado\n",
    "    a.max_var_iterations = 2000 #*500\n",
    "    a.md_level = refine.slow # Nivel del rifinamiento\n",
    "    a.repeat_optimization = 3 ##Originalmente comentado\n",
    "    \n",
    "    # Más info sobre el refinamiento: https://salilab.org/modeller/9.21/manual/node19.html\n",
    "    a.make()\n",
    "\n",
    "    ###########################\n",
    "    # Renombra el archivo pdb a code + _full.pdb\n",
    "    # DEBE ser el único archivo pdb en el directorio '.'\n",
    "\n",
    "    model_file = glob.glob('./' + code + '*.pdb')[0] # Nombre del archivo pdb\n",
    "    os.rename(model_file, pdbs_model_dir + code + \"_full.pdb\") # Mueve el pdb a la carpeta model_pdbs\n",
    "\n",
    "    # Elimina los archivos extra generados - Todos empiezan con el codigo del PDB ID\n",
    "    for f in glob.glob(code + \"*\"):\n",
    "        os.remove(f)\n",
    "\n",
    "end = time. time()\n",
    "print(end - start)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Corrección de modelos"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# PENDIENTE: Modeller por pasos, por comentar\n",
    "\n",
    "## Gaps, covertura e identidad \n",
    "#### Para los metadatos de los cristales obtengo una tabla con el número de gaps, la covertura de la secuencia y su identidad.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Hacer el alineamiento y crear el archivo de entrada para modeller\n",
    "Se lleva a cabo un alineamiento global entre la secuencia de UniProt (la cual está completa) y la secuencia de la estructura del cristal, la cual puede tener gaps. El alineamiento es global y tiene una alta penalización  a la apertura de nuevos gaps, lo cual reduce el número posible de gaps concordando con las estructuras cristalográficas."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Alineamiento\n",
    "from Bio import pairwise2\n",
    "\n",
    "# Primer alineamiento global, se penalizan con -10 los gaps abiertos\n",
    "alignment = pairwise2.align.localxs(seq_cry, cdk2_P24941, -10, -0.5, \n",
    "                                    gap_char='-')[0] \n",
    "algn1_struc = alignment[0] # Estructura\n",
    "algn2_seq = alignment[1]   # Secuencia UniProt\n",
    "\n",
    "seq_id = \"P24941\"\n",
    "cry_code = \"4FKW\"\n",
    "\n",
    "# Necesario: There should be 10 fields separated by colons, :\n",
    "# Please check the file to make sure your sequences end with the '*' character.\n",
    "# Nomenclaturas de los campos del header: https://salilab.org/modeller/8v2/manual/node176.html\n",
    "\n",
    "# Headers\n",
    "struc_header = \"structureX:\" + cry_code + \":.:\" + chaid + \":.:\" + chaid + \":.:.:.:\"\n",
    "seq_header = \"sequence:.:.:.:.:.:.:.:.:\"\n",
    "\n",
    "# PDB name / nombre del archivo que tendrá el pdb al final del modelado\n",
    "tail_pdb = \".pdb_chA.pdb\"\n",
    "\n",
    "# Crea el archivo de alineamiento con la estructura requerida por modeller\n",
    "filename = \"alignment.fasta\"\n",
    "with open(filename, \"w\") as handle:\n",
    "    handle.write(\"\\n>P1;%s%s\\n%s\\n%s*\\n>P1;%s\\n%s\\n%s*\\n\" % (cry_code, tail_pdb, struc_header, algn1_struc, \n",
    "                                                           seq_id, seq_header, algn2_seq))"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "# from modeller import *\n",
    "# Get the sequence of the 1qg8 PDB file, and write to an alignment file\n",
    "code = '../PDBs_chains/4FKW.pdb_chA.pdb'\n",
    "\n",
    "e = environ()\n",
    "m = model(e, file=code)\n",
    "aln = alignment(e)\n",
    "aln.append_model(m, align_codes=code)\n",
    "aln.write(file=\"P_1\"+'.seq')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Función para identificar los gaps\n",
    "Los gap se identifican en la secuencia de la estructura cristalográfica."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'num_gaps': 2,\n",
       " 'gap_lengths': [5, 13],\n",
       " 'gap_list': [[38, 42], [150, 162]],\n",
       " 'gap_window': [[37, 43], [149, 163]]}"
      ]
     },
     "execution_count": 51,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "### Encontrar el número e inicio y final de los gaps en el alineamiento\n",
    "import re\n",
    "\n",
    "def find_gaps(seq, r=1): # r es el número de residuos móviles al lado de la ventana del gap\n",
    "    gaps = re.findall('[-]+', seq) # todas las subsecuencias que tienen uno o más guiones\n",
    "    num_gaps = len(gaps); gap_lengths = []\n",
    "    gap_list = []; gap_window = []\n",
    "    # Obtener el índice de inicio y final del gap\n",
    "    for i , gap in enumerate(gaps,1):\n",
    "        start = seq.index(gap) + 1\n",
    "        end = start + len(gap) - 1\n",
    "        gap_lengths.append(len(gap))\n",
    "        gap_list.append([start, end])\n",
    "        gap_window.append([start - r, end + r])\n",
    "    return({\"num_gaps\": num_gaps, \"gap_lengths\":gap_lengths,\n",
    "            \"gap_list\": gap_list, \"gap_window\": gap_window})\n",
    "\n",
    "find_gaps(algn1_struc)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "self.residue_range('38', '42'), self.residue_range('150', '162')\n"
     ]
    }
   ],
   "source": [
    "# Preguntar si hay al menos un gap\n",
    "# si sí, crear para el primero\n",
    "\n",
    "gaps = find_gaps(algn1_struc)\n",
    "num_gaps = gaps[\"num_gaps\"]\n",
    "gap_i = gaps[\"gap_list\"]\n",
    "s = \"self.residue_range('\" + str(gap_i[0][0]) + \"', '\" + str(gap_i[0][1]) + \"')\"\n",
    "\n",
    "# A pesar de que sólo haya una, se ejecuta el ciclo, si hay más de una, se crean para las demás\n",
    "for i in range(1, num_gaps):\n",
    "    s = s + \", \" + \"self.residue_range('\" + str(gap_i[i][0]) + \"', '\" + str(gap_i[i][1]) + \"')\" #nótese la coma\n",
    "    \n",
    "print(s)\n",
    "    \n",
    "# k = [exec(\"%s('37', '43')\" % (x))]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 85,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "check_ali___> Checking the sequence-structure alignment. \n",
      "\n",
      "Implied intrachain target CA(i)-CA(i+1) distances longer than  8.0 angstroms:\n",
      "\n",
      "ALN_POS  TMPL  RID1  RID2  NAM1  NAM2     DIST\n",
      "----------------------------------------------\n",
      "END OF TABLE\n",
      "read_to_681_> topology.submodel read from topology file:        3\n",
      "mdtrsr__446W> A potential that relies on one protein is used, yet you have at\n",
      "              least one known structure available. MDT, not library, potential is used.\n",
      "0 atoms in HETATM/BLK residues constrained\n",
      "to protein atoms within 2.30 angstroms\n",
      "and protein CA atoms within 10.00 angstroms\n",
      "0 atoms in residues without defined topology\n",
      "constrained to be rigid bodies\n",
      "condens_443_> Restraints marked for deletion were removed.\n",
      "              Total number of restraints before, now:    30219    28030\n",
      "148 (of 2398 total) atoms selected for optimization\n",
      "iupac_m_397W> Atoms were not swapped because of the uncertainty of how to handle the H atom.\n",
      "iupac_m_397W> Atoms were not swapped because of the uncertainty of how to handle the H atom.\n",
      "iupac_m_397W> Atoms were not swapped because of the uncertainty of how to handle the H atom.\n",
      "\n",
      "\n",
      ">> ENERGY; Differences between the model's features and restraints:\n",
      "Number of all residues in MODEL                   :      298\n",
      "Number of all, selected real atoms                :     2398     148\n",
      "Number of all, selected pseudo atoms              :        0       0\n",
      "Number of all static, selected restraints         :    28030     619\n",
      "COVALENT_CYS                                      :        F\n",
      "NONBONDED_SEL_ATOMS                               :        1\n",
      "Number of non-bonded pairs (excluding 1-2,1-3,1-4):      190\n",
      "Dynamic pairs routine                             : 5, NATM_SEL x NATM double loop\n",
      "Atomic shift for contacts update (UPDATE_DYNAMIC) :    0.390\n",
      "LENNARD_JONES_SWITCH                              :    6.500   7.500\n",
      "COULOMB_JONES_SWITCH                              :    6.500   7.500\n",
      "RESIDUE_SPAN_RANGE                                :        0   99999\n",
      "NLOGN_USE                                         :       15\n",
      "CONTACT_SHELL                                     :    4.000\n",
      "DYNAMIC_PAIRS,_SPHERE,_COULOMB,_LENNARD,_MODELLER :        T       T       F       F       F\n",
      "SPHERE_STDV                                       :    0.050\n",
      "RADII_FACTOR                                      :    0.820\n",
      "Current energy                                    :          93.3959\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Summary of the restraint violations: \n",
      "\n",
      "   NUM     ... number of restraints.\n",
      "   NUMVI   ... number of restraints with RVIOL > VIOL_REPORT_CUT[i].\n",
      "   RVIOL   ... relative difference from the best value.\n",
      "   NUMVP   ... number of restraints with -Ln(pdf) > VIOL_REPORT_CUT2[i].\n",
      "   RMS_1   ... RMS(feature, minimally_violated_basis_restraint, NUMB).\n",
      "   RMS_2   ... RMS(feature, best_value, NUMB).\n",
      "   MOL.PDF ... scaled contribution to -Ln(Molecular pdf).\n",
      "\n",
      " #                     RESTRAINT_GROUP      NUM   NUMVI  NUMVP   RMS_1   RMS_2         MOL.PDF     S_i\n",
      "------------------------------------------------------------------------------------------------------\n",
      " 1 Bond length potential              :     154       0      0   0.003   0.003     0.32737       1.000\n",
      " 2 Bond angle potential               :     214       0      0   1.475   1.475      9.0913       1.000\n",
      " 3 Stereochemical cosine torsion poten:     106       0      5  46.984  46.984      37.130       1.000\n",
      " 4 Stereochemical improper torsion pot:      69       0      0   0.979   0.979      1.6051       1.000\n",
      " 5 Soft-sphere overlap restraints     :     190       0      0   0.002   0.002     0.65640E-01   1.000\n",
      " 6 Lennard-Jones 6-12 potential       :       0       0      0   0.000   0.000      0.0000       1.000\n",
      " 7 Coulomb point-point electrostatic p:       0       0      0   0.000   0.000      0.0000       1.000\n",
      " 8 H-bonding potential                :       0       0      0   0.000   0.000      0.0000       1.000\n",
      " 9 Distance restraints 1 (CA-CA)      :       0       0      0   0.000   0.000      0.0000       1.000\n",
      "10 Distance restraints 2 (N-O)        :       0       0      0   0.000   0.000      0.0000       1.000\n",
      "11 Mainchain Phi dihedral restraints  :       0       0      0   0.000   0.000      0.0000       1.000\n",
      "12 Mainchain Psi dihedral restraints  :       0       0      0   0.000   0.000      0.0000       1.000\n",
      "13 Mainchain Omega dihedral restraints:      20       0      0   4.221   4.221      3.4752       1.000\n",
      "14 Sidechain Chi_1 dihedral restraints:      16       0      0  81.387  81.387      4.4588       1.000\n",
      "15 Sidechain Chi_2 dihedral restraints:      10       0      0  81.214  81.214      4.9987       1.000\n",
      "16 Sidechain Chi_3 dihedral restraints:       6       0      0 117.063 117.063      4.3443       1.000\n",
      "17 Sidechain Chi_4 dihedral restraints:       2       0      0 132.185 132.185      1.4303       1.000\n",
      "18 Disulfide distance restraints      :       0       0      0   0.000   0.000      0.0000       1.000\n",
      "19 Disulfide angle restraints         :       0       0      0   0.000   0.000      0.0000       1.000\n",
      "20 Disulfide dihedral angle restraints:       0       0      0   0.000   0.000      0.0000       1.000\n",
      "21 Lower bound distance restraints    :       0       0      0   0.000   0.000      0.0000       1.000\n",
      "22 Upper bound distance restraints    :       0       0      0   0.000   0.000      0.0000       1.000\n",
      "23 Distance restraints 3 (SDCH-MNCH)  :       0       0      0   0.000   0.000      0.0000       1.000\n",
      "24 Sidechain Chi_5 dihedral restraints:       0       0      0   0.000   0.000      0.0000       1.000\n",
      "25 Phi/Psi pair of dihedral restraints:      22      14      6  36.905 128.248      26.470       1.000\n",
      "26 Distance restraints 4 (SDCH-SDCH)  :       0       0      0   0.000   0.000      0.0000       1.000\n",
      "27 Distance restraints 5 (X-Y)        :       0       0      0   0.000   0.000      0.0000       1.000\n",
      "28 NMR distance restraints 6 (X-Y)    :       0       0      0   0.000   0.000      0.0000       1.000\n",
      "29 NMR distance restraints 7 (X-Y)    :       0       0      0   0.000   0.000      0.0000       1.000\n",
      "30 Minimal distance restraints        :       0       0      0   0.000   0.000      0.0000       1.000\n",
      "31 Non-bonded restraints              :       0       0      0   0.000   0.000      0.0000       1.000\n",
      "32 Atomic accessibility restraints    :       0       0      0   0.000   0.000      0.0000       1.000\n",
      "33 Atomic density restraints          :       0       0      0   0.000   0.000      0.0000       1.000\n",
      "34 Absolute position restraints       :       0       0      0   0.000   0.000      0.0000       1.000\n",
      "35 Dihedral angle difference restraint:       0       0      0   0.000   0.000      0.0000       1.000\n",
      "36 GBSA implicit solvent potential    :       0       0      0   0.000   0.000      0.0000       1.000\n",
      "37 EM density fitting potential       :       0       0      0   0.000   0.000      0.0000       1.000\n",
      "38 SAXS restraints                    :       0       0      0   0.000   0.000      0.0000       1.000\n",
      "39 Symmetry restraints                :       0       0      0   0.000   0.000      0.0000       1.000\n",
      "\n",
      "\n",
      "\n",
      "# Heavy relative violation of each residue is written to: 4FKW_full.V99990001\n",
      "# The profile is NOT normalized by the number of restraints.\n",
      "# The profiles are smoothed over a window of residues:    1\n",
      "# The sum of all numbers in the file:    3651.7388\n",
      "\n",
      "\n",
      "\n",
      "List of the violated restraints:\n",
      "   A restraint is violated when the relative difference\n",
      "   from the best value (RVIOL) is larger than CUTOFF.\n",
      "\n",
      "   ICSR   ... index of a restraint in the current set.\n",
      "   RESNO  ... residue numbers of the first two atoms.\n",
      "   ATM    ... IUPAC atom names of the first two atoms.\n",
      "   FEAT   ... the value of the feature in the model.\n",
      "   restr  ... the mean of the basis restraint with the smallest\n",
      "              difference from the model (local minimum).\n",
      "   viol   ... difference from the local minimum.\n",
      "   rviol  ... relative difference from the local minimum.\n",
      "   RESTR  ... the best value (global minimum).\n",
      "   VIOL   ... difference from the best value.\n",
      "   RVIOL  ... relative difference from the best value.\n",
      "\n",
      "\n",
      "-------------------------------------------------------------------------------------------------\n",
      "\n",
      "Feature 25                           : Phi/Psi pair of dihedral restraints     \n",
      "List of the RVIOL violations larger than   :       6.5000\n",
      "\n",
      "    #   ICSR  RESNO1/2 ATM1/2   INDATM1/2    FEAT   restr    viol   rviol   RESTR    VIOL   RVIOL\n",
      "    1   8490  37L  38D C   N     294  296 -161.12  -96.50   67.31    2.74  -63.30  167.01   17.97\n",
      "    1         38D  38D N   CA    296  297   95.36  114.20                  -40.00\n",
      "    2   8491  38D  39T C   N     302  304 -144.89 -124.80   43.12    1.36  -63.20  158.86   25.59\n",
      "    2         39T  39T N   CA    304  305 -178.34  143.50                  -42.10\n",
      "    3   8493  40E  41T C   N     318  320 -101.93  -78.10   50.19    1.64  -63.20  129.84   19.23\n",
      "    3         41T  41T N   CA    320  321 -166.03  149.80                  -42.10\n",
      "    4   8494  41T  42E C   N     325  327 -110.67 -117.80   21.21    0.90  -63.60  164.03   19.68\n",
      "    4         42E  42E N   CA    327  328  116.83  136.80                  -40.30\n",
      "    5   8495  42E  43G C   N     334  336  161.11 -167.20   39.28    1.36  -80.20  120.84    8.61\n",
      "    5         43G  43G N   CA    336  337  151.40  174.60                  174.10\n",
      "    6   8602 149A 150R C   N    1187 1189 -147.87 -125.20   57.26    2.19  -63.00  151.68   26.10\n",
      "    6        150R 150R N   CA   1189 1190 -166.82  140.60                  -41.10\n",
      "    7   8603 150R 151A C   N    1198 1200  -71.41  -68.20   34.50    2.89  -62.50  152.11   24.51\n",
      "    7        151A 151A N   CA   1200 1201  110.96  145.30                  -40.90\n",
      "    8   8604 151A 152F C   N    1203 1205 -143.75 -124.20   29.79    0.83  -63.20  170.19   28.76\n",
      "    8        152F 152F N   CA   1205 1206  165.78  143.30                  -44.30\n",
      "    9   8605 152F 153G C   N    1214 1216 -166.47 -167.20    4.74    0.18   82.20 -163.92   14.33\n",
      "    9        153G 153G N   CA   1216 1217  169.92  174.60                    8.50\n",
      "   10   8609 156V 157R C   N    1239 1241 -115.17 -125.20   29.86    1.19  -63.00  162.19   19.01\n",
      "   10        157R 157R N   CA   1241 1242  112.47  140.60                  -41.10\n",
      "   11   8611 158T 159Y C   N    1257 1259 -118.88 -124.30    8.49    0.67  -63.50 -176.77   31.88\n",
      "   11        159Y 159Y N   CA   1259 1260  141.94  135.40                  -43.40\n",
      "   12   8612 159Y 160T C   N    1269 1271  -84.66  -78.10   14.44    0.47  -63.20  156.72   21.53\n",
      "   12        160T 160T N   CA   1271 1272  162.66  149.80                  -42.10\n",
      "   13   8613 160T 161H C   N    1276 1278 -142.07 -125.60   96.13    3.32  -63.20  115.36   19.24\n",
      "   13        161H 161H N   CA   1278 1279 -126.49  138.80                  -42.30\n",
      "   14   8614 161H 162E C   N    1286 1288  -62.25  -69.30    7.85    0.70  -63.60  173.75   23.09\n",
      "   14        162E 162E N   CA   1288 1289  145.95  142.50                  -40.30\n",
      "\n",
      "\n",
      "report______> Distribution of short non-bonded contacts:\n",
      "\n",
      "  serious non-bonded atom clash:   115 1180   2.258\n",
      "  serious non-bonded atom clash:   116 1180   1.308\n",
      "  serious non-bonded atom clash:   117  274   1.922\n",
      "  serious non-bonded atom clash:   118  273   2.250\n",
      "  serious non-bonded atom clash:   118 1180   1.811\n",
      "  serious non-bonded atom clash:   119  271   2.120\n",
      "  serious non-bonded atom clash:   119  272   1.776\n",
      "  serious non-bonded atom clash:   119  273   2.083\n",
      "  serious non-bonded atom clash:   119  274   1.334\n",
      "  serious non-bonded atom clash:   120  271   2.287\n",
      "  serious non-bonded atom clash:   120  273   1.444\n",
      "  serious non-bonded atom clash:   121  271   2.146\n",
      "  serious non-bonded atom clash:   121  273   1.635\n",
      "\n",
      "DISTANCE1:  0.00 2.10 2.20 2.30 2.40 2.50 2.60 2.70 2.80 2.90 3.00 3.10 3.20 3.30 3.40\n",
      "DISTANCE2:  2.10 2.20 2.30 2.40 2.50 2.60 2.70 2.80 2.90 3.00 3.10 3.20 3.30 3.40 3.50\n",
      "FREQUENCY:     8    2    3    2    0    6    6   32   81  125  132  178  245  339  358\n",
      "\n",
      "\n",
      "<< end of ENERGY.\n",
      "\n",
      ">> Summary of successfully produced models:\n",
      "Filename                          molpdf\n",
      "----------------------------------------\n",
      "4FKW_full.B99990001.pdb         93.39586\n",
      "\n"
     ]
    }
   ],
   "source": [
    "from modeller import *\n",
    "from modeller.automodel import *\n",
    "\n",
    "# log.verbose()\n",
    "env = environ()\n",
    "\n",
    "env.io.atom_files_directory = ['.', '../PDBs_chains/']\n",
    "\n",
    "\n",
    "# Modifico loopmodel para fijar los residuos ya existentes en el cristal\n",
    "# NECESARIO para poder capturar los valores obtenidos por el rango de gaps\n",
    "code = \"\"\"\n",
    "class MyModel(automodel):\n",
    "    def select_atoms(self):\n",
    "        return selection(\"\"\" + s + \"\"\")\n",
    "\"\"\"\n",
    "\n",
    "exec(code)\n",
    "\n",
    "a = MyModel(env, alnfile = 'alignment.fasta',\n",
    "              knowns = \"4FKW.pdb_chA.pdb\", sequence = seq_id)\n",
    "\n",
    "a.starting_model= 1\n",
    "a.ending_model  = 1\n",
    "\n",
    "\n",
    "a.make()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
