{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "hideCode": false,
    "hidePrompt": false
   },
   "outputs": [],
   "source": [
    "# Cargar cada estructura desde el archivo que proporciona DUD"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 110,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import glob\n",
    "from prody import *\n",
    "path_to_dud_target_files = os.path.join(*'../../../Benchmarks/dud_targets'.split('/'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio.Seq import Seq\n",
    "from Bio.Blast import NCBIXML\n",
    "from Bio.Blast.Applications import NcbiblastpCommandline\n",
    "\n",
    "def get_uniprot_id_from_pdb_structure(prot_name, verbose = True,\n",
    "                                      path_to_files = F'{path_to_dud_target_files}/'):\n",
    "    # Read the .pdb file and obtains its sequence\n",
    "    prot_name = prot_name.lower()\n",
    "    rec_pdb_file = glob.glob(F'{path_to_files}/{prot_name}/*pdb')[0]\n",
    "    rec_pdb = parsePDB(rec_pdb_file)\n",
    "    resnums = rec_pdb.select('ca').getResnums()\n",
    "    resids = rec_pdb.select('ca').getSequence()\n",
    "    positions = dict(zip(resnums, resids))\n",
    "    sequence = ''\n",
    "    \n",
    "    for i in range(1, resnums[-1]):\n",
    "        try:\n",
    "            sequence += positions[i]\n",
    "        except:\n",
    "            sequence += '-'\n",
    "    while sequence[:5] == '-----':\n",
    "        sequence = sequence[5:]\n",
    "\n",
    "    # Saves the sequence  in a temporal fasta file\n",
    "    with open(\"./temp_fasta.fasta\", \"w\") as ofile:\n",
    "        ofile.write(\">\" + 'Seq' + \"\\n\" + sequence + \"\\n\")\n",
    "        ofile.close()\n",
    "    # Perfroms the blastp over uniprot swissprot\n",
    "    blastx_cline = NcbiblastpCommandline(cmd='blastp', query='temp_fasta.fasta', \n",
    "                    db = os.path.join(*'../../../Benchmarks/uniprot_db/uniprot_sprot.fasta'.split('/')), \n",
    "                    evalue = 0.0001, outfmt = 5, out=\"blast_results.xml\")\n",
    "    blastx_cline()\n",
    "    # Reads the temporary results\n",
    "    blast_record = NCBIXML.read(open(\"blast_results.xml\"))\n",
    "    # Gets the first record, with the highest score and from HUMAN\n",
    "    try:\n",
    "        human_titles = [i.title for i in blast_record.alignments if 'human' in i.title.lower()]\n",
    "        title = human_titles[0]\n",
    "        uniprot_id = title.split('|')[3]\n",
    "        name = title.split('|')[4]\n",
    "    except:\n",
    "        title = blast_record.alignments[0].title\n",
    "        uniprot_id = blast_record.alignments[0].title.split('|')[3]\n",
    "        name = title.split('|')[4] + '*'\n",
    "    if verbose:\n",
    "        print(prot_name, uniprot_id, '\\n', title)\n",
    "    return {'name': name, 'accession': uniprot_id}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 108,
   "metadata": {},
   "outputs": [],
   "source": [
    "# First we have to obtain the UNIPROT Accession numbers\n",
    "# dud_targets_uniprot = {}\n",
    "# for name in dud_targets_names.keys():\n",
    "#     dud_targets_uniprot[name] = get_uniprot_id_from_pdb_structure(prot_name = name)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Saving the result"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Saving the results\n",
    "import json\n",
    "# with open('../../../Benchmarks/dud_targets_uniprot_accessions.json', 'w') as fp:\n",
    "#     json.dump(dud_targets_uniprot, fp)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 114,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Targets from DEKOIS2.0"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 90,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pypdb\n",
    "import time"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 115,
   "metadata": {},
   "outputs": [],
   "source": [
    "# dekois_targets_uniprot = {}\n",
    "# for key, pdb_id in dekois_targets_names.items():\n",
    "#     dict_info = pypdb.get_all_info(pdb_id)['polymer']\n",
    "    \n",
    "#     if type(dict_info) is list:\n",
    "#         molecule = dict_info[0]['macroMolecule']\n",
    "#     elif type(dict_info) is dict:\n",
    "#         molecule = dict_info['macroMolecule']\n",
    "#     if type(molecule) is list:\n",
    "#         molecule = molecule[0]\n",
    "#     name = molecule['@name']\n",
    "#     uniprot_id = molecule['accession']['@id']\n",
    "#     dekois_targets_uniprot[key] = {'name': name, 'accession': uniprot_id}\n",
    "#     time.sleep(0.5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 107,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Saving the results\n",
    "# with open('../../../Benchmarks/dekois2_targets_uniprot_accessions.json', 'w') as fp:\n",
    "#     json.dump(dekois_targets_uniprot, fp)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Getting the number of crystals in pdb"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 294,
   "metadata": {},
   "outputs": [],
   "source": [
    "# opening the results\n",
    "\n",
    "dekois_accessions = pd.read_json('../../../Benchmarks/dekois2_targets_uniprot_accessions.json').T\n",
    "dud_accessions = pd.read_json('../../../Benchmarks/dud_targets_uniprot_accessions.json').T"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 295,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of targets in DUD set: 21\n",
      "Number of targets in DEKOIS2.0 set: 81\n"
     ]
    }
   ],
   "source": [
    "# Number of targets on each set\n",
    "print(F'Number of targets in DUD set: {dud_accessions.shape[0]}')\n",
    "\n",
    "print(F'Number of targets in DEKOIS2.0 set: {dekois_accessions.shape[0]}')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Then we have to merge the dataframes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 304,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['O14965',\n",
       " 'O15530',\n",
       " 'O76074',\n",
       " 'P00374',\n",
       " 'P00374',\n",
       " 'P00533',\n",
       " 'P00533',\n",
       " 'P00720',\n",
       " 'P00720',\n",
       " 'P00734',\n",
       " 'P00742',\n",
       " 'P00742',\n",
       " 'P00749',\n",
       " 'P00750',\n",
       " 'P00811',\n",
       " 'P00813',\n",
       " 'P03366',\n",
       " 'P03366',\n",
       " 'P03372',\n",
       " 'P03372',\n",
       " 'P04035',\n",
       " 'P04058',\n",
       " 'P04062',\n",
       " 'P04150',\n",
       " 'P04150',\n",
       " 'P04183',\n",
       " 'P04585',\n",
       " 'P04587',\n",
       " 'P04626',\n",
       " 'P04818',\n",
       " 'P05979',\n",
       " 'P06239',\n",
       " 'P06401',\n",
       " 'P06737',\n",
       " 'P06737',\n",
       " 'P07711',\n",
       " 'P07900',\n",
       " 'P08069',\n",
       " 'P08253',\n",
       " 'P09874',\n",
       " 'P0C6X7',\n",
       " 'P10275',\n",
       " 'P10275',\n",
       " 'P10415',\n",
       " 'P11217',\n",
       " 'P11309',\n",
       " 'P11362',\n",
       " 'P11362',\n",
       " 'P11509',\n",
       " 'P12821',\n",
       " 'P12821',\n",
       " 'P12931',\n",
       " 'P14061',\n",
       " 'P15056',\n",
       " 'P15121',\n",
       " 'P17948',\n",
       " 'P19793',\n",
       " 'P19793',\n",
       " 'P19971',\n",
       " 'P21964',\n",
       " 'P22102',\n",
       " 'P22303',\n",
       " 'P23219',\n",
       " 'P24941',\n",
       " 'P24941',\n",
       " 'P27907',\n",
       " 'P28845',\n",
       " 'P31749',\n",
       " 'P35354',\n",
       " 'P35968',\n",
       " 'P43235',\n",
       " 'P45983',\n",
       " 'P45984',\n",
       " 'P48736',\n",
       " 'P49137',\n",
       " 'P49841',\n",
       " 'P52333',\n",
       " 'P52732',\n",
       " 'P53779',\n",
       " 'P54760',\n",
       " 'P55859',\n",
       " 'P62942',\n",
       " 'P78536',\n",
       " 'P80276',\n",
       " 'P9WGR1',\n",
       " 'Q00987',\n",
       " 'Q02763',\n",
       " 'Q04759',\n",
       " 'Q05769',\n",
       " 'Q07343',\n",
       " 'Q07869',\n",
       " 'Q08881',\n",
       " 'Q13464',\n",
       " 'Q16539',\n",
       " 'Q16769',\n",
       " 'Q6DE08',\n",
       " 'Q8IXJ6',\n",
       " 'Q92731',\n",
       " 'Q92769',\n",
       " 'Q9BY41',\n",
       " 'Q9BYF1',\n",
       " 'Q9P1W9']"
      ]
     },
     "execution_count": 304,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# First we create a set of uniprot_ids\n",
    "acc_num_set = (dud_accessions['accession'].to_list() +\n",
    "                  dekois_accessions['accession'].to_list())\n",
    "len(acc_num_set)\n",
    "sorted(acc_num_set)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 246,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>accession</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>11-beta-HSD1</th>\n",
       "      <td>P28845</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17-beta-HSD1</th>\n",
       "      <td>P14061</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A2A</th>\n",
       "      <td>P00720</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ACE</th>\n",
       "      <td>P12821</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ACE-2</th>\n",
       "      <td>Q9BYF1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TPA</th>\n",
       "      <td>P00750</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TS</th>\n",
       "      <td>P04818</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>uPA</th>\n",
       "      <td>P00749</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>VEGFR1</th>\n",
       "      <td>P17948</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>VEGFR2</th>\n",
       "      <td>P35968</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>81 rows × 1 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             accession\n",
       "11-beta-HSD1    P28845\n",
       "17-beta-HSD1    P14061\n",
       "A2A             P00720\n",
       "ACE             P12821\n",
       "ACE-2           Q9BYF1\n",
       "...                ...\n",
       "TPA             P00750\n",
       "TS              P04818\n",
       "uPA             P00749\n",
       "VEGFR1          P17948\n",
       "VEGFR2          P35968\n",
       "\n",
       "[81 rows x 1 columns]"
      ]
     },
     "execution_count": 246,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dekois_accessions[['accession']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 354,
   "metadata": {},
   "outputs": [],
   "source": [
    "dekois_accessions = pd.read_json('../../../Benchmarks/dekois2_targets_uniprot_accessions.json').T\n",
    "dud_accessions = pd.read_json('../../../Benchmarks/dud_targets_uniprot_accessions.json').T\n",
    "\n",
    "# Reset the index\n",
    "dud_accessions.reset_index(inplace=True)\n",
    "dekois_accessions.reset_index(inplace=True)\n",
    "\n",
    "# Setting the uniprot accession index as the new index\n",
    "dud_accessions.set_index('accession', inplace = True)\n",
    "dekois_accessions.set_index('accession', inplace = True)\n",
    "\n",
    "# Merging the datasets\n",
    "df_targets = pd.concat([a,b], axis = 0).sort_index()\n",
    "df_targets = c.iloc[~c.index.duplicated(keep='last')]\n",
    "# Naming the table\n",
    "df_targets.columns = ['name', 'long_name']\n",
    "\n",
    "# Fill a new columns indicating if the target is in a given datset\n",
    "# for dud\n",
    "df_targets.loc[:, 'DUD'] = [1 if i in dud_accessions.index else 0 for i in df_targets.index]\n",
    "# for dekois\n",
    "df_targets.loc[:, 'DEKOIS2'] = [1 if i in dekois_accessions.index else 0 for i in df_targets.index]\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 355,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>name</th>\n",
       "      <th>long_name</th>\n",
       "      <th>DUD</th>\n",
       "      <th>DEKOIS2</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>accession</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>O14965</th>\n",
       "      <td>AURKA</td>\n",
       "      <td>Aurora kinase A</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>O15530</th>\n",
       "      <td>PDK1</td>\n",
       "      <td>3-phosphoinositide-dependent protein kinase 1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>O76074</th>\n",
       "      <td>PDE5</td>\n",
       "      <td>cGMP-specific 3',5'-cyclic phosphodiesterase</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P00374</th>\n",
       "      <td>DHFR</td>\n",
       "      <td>Dihydrofolate reductase</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P00533</th>\n",
       "      <td>EGFR</td>\n",
       "      <td>Epidermal growth factor receptor</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Q92731</th>\n",
       "      <td>ER-beta</td>\n",
       "      <td>Estrogen receptor beta</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Q92769</th>\n",
       "      <td>HDAC2</td>\n",
       "      <td>Histone deacetylase 2</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Q9BY41</th>\n",
       "      <td>HDAC8</td>\n",
       "      <td>Histone deacetylase 8</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Q9BYF1</th>\n",
       "      <td>ACE-2</td>\n",
       "      <td>Angiotensin-converting enzyme 2</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Q9P1W9</th>\n",
       "      <td>PIM-2</td>\n",
       "      <td>Serine/threonine-protein kinase pim-2</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>89 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "              name                                      long_name  DUD  \\\n",
       "accession                                                                \n",
       "O14965       AURKA                                Aurora kinase A    0   \n",
       "O15530        PDK1  3-phosphoinositide-dependent protein kinase 1    0   \n",
       "O76074        PDE5   cGMP-specific 3',5'-cyclic phosphodiesterase    0   \n",
       "P00374        DHFR                        Dihydrofolate reductase    1   \n",
       "P00533        EGFR               Epidermal growth factor receptor    1   \n",
       "...            ...                                            ...  ...   \n",
       "Q92731     ER-beta                         Estrogen receptor beta    0   \n",
       "Q92769       HDAC2                          Histone deacetylase 2    0   \n",
       "Q9BY41       HDAC8                          Histone deacetylase 8    0   \n",
       "Q9BYF1       ACE-2                Angiotensin-converting enzyme 2    0   \n",
       "Q9P1W9       PIM-2          Serine/threonine-protein kinase pim-2    0   \n",
       "\n",
       "           DEKOIS2  \n",
       "accession           \n",
       "O14965           1  \n",
       "O15530           1  \n",
       "O76074           1  \n",
       "P00374           1  \n",
       "P00533           1  \n",
       "...            ...  \n",
       "Q92731           1  \n",
       "Q92769           1  \n",
       "Q9BY41           1  \n",
       "Q9BYF1           1  \n",
       "Q9P1W9           1  \n",
       "\n",
       "[89 rows x 4 columns]"
      ]
     },
     "execution_count": 355,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_targets"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "hideCode": false,
    "hideOutput": true,
    "hidePrompt": false
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {
    "hideCode": false,
    "hidePrompt": false
   },
   "outputs": [],
   "source": [
    "dud_targets_names = {\n",
    "    \"ACE\": \"Angiotensin-converting enzyme\",\n",
    "    \"ACHE\": \"Acetylcholine esterase\",\n",
    "    \"ADA\": \"Adenosine deaminase\",\n",
    "    \"ALR2\": \"Aldose reductase\",\n",
    "    \"AmpC\": \"AmpC beta lactamase\",\n",
    "    \"AR\": \"Androgen receptor\",\n",
    "    \"CDK2\": \"Cyclin dependent kinase 2\",\n",
    "    \"COMT\": \"Catechol O-methyltransferase\",\n",
    "    \"COX1\": \"Cyclooxygenase 1\",\n",
    "    \"COX2\": \"Cyclooxygenase 2\",\n",
    "    \"DHFR\": \"Dihydrofolate reductase\",\n",
    "    \"EGFr\": \"Epidermal growth factor receptor kinase\",\n",
    "    \"ER_agonist\": \"Estrogen receptor agonist\",\n",
    "    \"ER_antagonist\": \"Estrogen receptor antagonist\",\n",
    "    \"FGFr1\": \"Fibroblast growth factor receptor kinase\",\n",
    "    \"FXa\": \"Factor Xa\",\n",
    "    \"GART\": \"glycinamide ribonucleotide transformylase\",\n",
    "    \"GPB\": \"Glycogen phosphorylase beta\",\n",
    "    \"GR\": \"Glutocorticoid receptor\",\n",
    "    \"HIVPR\": \"HIV protease\",\n",
    "    \"HIVRT\": \"HIV reverse transcriptase\",\n",
    "    \"HMGR\": \"Hydroxymethylglutaryl-CoA reductase\",\n",
    "    \"HSP90\": \"Human heat shock protein 90 kinase\",\n",
    "    \"InhA\": \"Enoyl ACP reductase\",\n",
    "    \"MR\": \"Mineralcorticoid receptor\",\n",
    "    \"NA\": \"Neuraminidase\",\n",
    "    \"P38\": \"P38 mitogen activated protein kinase\",\n",
    "    \"PARP\": \"Poly(ADP-ribose) polymerase\",\n",
    "    \"PDE5\": \"Phosphodiesterase V\",\n",
    "    \"PDGFrb\": \"Platlet derived growth factor receptor kinase\",\n",
    "    \"PNP\": \"Purine nucleoside phosphorylase\",\n",
    "    \"PPARg\": \"Peroxisome proliferator activated receptor gamma\",\n",
    "    \"PR\": \"Progesterone receptor\",\n",
    "    \"RXRa\": \"Retinoic X receptor alpha\",\n",
    "    \"SAHH\": \"S-adenosyl-homocysteine hydrolase\",\n",
    "    \"SRC\": \"Tyrosine kinase SRC\",\n",
    "    \"Thrombin\": \"Thrombin\",\n",
    "    \"TK\": \"Thymidine kinase\",\n",
    "    \"Trypsin\": \"Trypsin\",\n",
    "    \"VEGFr2\": \"Vascular endothelial growth factor receptor kinase\"\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "dekois_targets_names = {\n",
    "    \"11-beta-HSD1\": \"3tfq\",\n",
    "    \"17-beta-HSD1\": \"3klm\",\n",
    "    \"A2A\": \"3eml\",\n",
    "    \"ACE\": \"1uze\",\n",
    "    \"ACE-2\": \"1r4l\",\n",
    "    \"ACHE\": \"1eve\",\n",
    "    \"ADAM17\": \"3ewj\",\n",
    "    \"ADRB2\": \"3ny9\",\n",
    "    \"AKT1\": \"3qkl\",\n",
    "    \"ALR2\": \"1ah3\",\n",
    "    \"AR\": \"1e3g\",\n",
    "    \"AURKA\": \"3fdn\",\n",
    "    \"AURKB\": \"2vgo\",\n",
    "    \"Bcl-2\": \"2w3l\",\n",
    "    \"BRAF\": \"3skc\",\n",
    "    \"CATL\": \"3bc3\",\n",
    "    \"CDK2\": \"1ckp\",\n",
    "    \"COX1\": \"3kk6\",\n",
    "    \"COX2\": \"1cx2\",\n",
    "    \"CTSK\": \"3kx1\",\n",
    "    \"CYP2A6\": \"1z11\",\n",
    "    \"DHFR\": \"1s3v\",\n",
    "    \"EGFR\": \"1m17\",\n",
    "    \"EPHB4\": \"2vwz\",\n",
    "    \"ERBB2\": \"3pp0\",\n",
    "    \"ER-beta\": \"3oll\",\n",
    "    \"FGFR1\": \"1agw\",\n",
    "    \"FKBP1A\": \"2dg3\",\n",
    "    \"FXA\": \"1f0r\",\n",
    "    \"GBA\": \"2wcg\",\n",
    "    \"GR\": \"1nhz\",\n",
    "    \"GSK3B\": \"3i4b\",\n",
    "    \"HDAC2\": \"3max\",\n",
    "    \"HDAC8\": \"3sff\",\n",
    "    \"HIV1PR\": \"3nu3\",\n",
    "    \"HIV1RT\": \"1s6p\",\n",
    "    \"HMGA\": \"1hw8\",\n",
    "    \"HSP90\": \"1uy6\",\n",
    "    \"IGF1R\": \"3nw7\",\n",
    "    \"INHA\": \"1p44\",\n",
    "    \"ITK\": \"3mj1\",\n",
    "    \"JAK3\": \"3lxl\",\n",
    "    \"JNK1\": \"3elj\",\n",
    "    \"JNK2\": \"3npc\",\n",
    "    \"JNK3\": \"2b1p\",\n",
    "    \"KIF11\": \"3k5e\",\n",
    "    \"Lck\": \"3mpm\",\n",
    "    \"MDM2\": \"3lbk\",\n",
    "    \"MK2\": \"3kc3\",\n",
    "    \"MMP2\": \"1hov\",\n",
    "    \"NA\": \"1a4g\",\n",
    "    \"P38\": \"1ouk\",\n",
    "    \"PARP-1\": \"3l3m\",\n",
    "    \"PDE4B\": \"3frg\",\n",
    "    \"PDE5\": \"1xp0\",\n",
    "    \"PDK1\": \"2xch\",\n",
    "    \"PI3Kg\": \"3dbs\",\n",
    "    \"PIM-1\": \"3r04\",\n",
    "    \"PIM-2\": \"2iwi\",\n",
    "    \"PNP\": \"1b8o\",\n",
    "    \"PPARa\": \"2p54\",\n",
    "    \"PPARg\": \"1fm9\",\n",
    "    \"PR\": \"2w8y\",\n",
    "    \"PRKCQ\": \"1xjd\",\n",
    "    \"PYGL-in\": \"1xoi\",\n",
    "    \"PYGL-out\": \"3dds\",\n",
    "    \"QPCT\": \"2afx\",\n",
    "    \"ROCK-1\": \"3v8s\",\n",
    "    \"RXR\": \"2p1t\",\n",
    "    \"SARS-HCoV\": \"2z94\",\n",
    "    \"SIRT2\": \"1j8f\",\n",
    "    \"SRC\": \"2src\",\n",
    "    \"Thrombin\": \"3rm2\",\n",
    "    \"TIE2\": \"2oo8\",\n",
    "    \"TK\": \"1w4r\",\n",
    "    \"TP\": \"1uou\",\n",
    "    \"TPA\": \"1a5h\",\n",
    "    \"TS\": \"1i00\",\n",
    "    \"uPA\": \"3mhw\",\n",
    "    \"VEGFR1\": \"3hng\",\n",
    "    \"VEGFR2\": \"3c7q\"\n",
    "}"
   ]
  }
 ],
 "metadata": {
  "hide_code_all_hidden": false,
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
