{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Protein targets"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "hideCode": false,
    "hidePrompt": false
   },
   "outputs": [],
   "source": [
    "# Cargar cada estructura desde el archivo que proporciona DUD"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "ename": "ModuleNotFoundError",
     "evalue": "No module named 'pypdb'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mModuleNotFoundError\u001b[0m                       Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-1-42cc5446cb00>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m      4\u001b[0m \u001b[0;32mimport\u001b[0m \u001b[0mglob\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      5\u001b[0m \u001b[0;32mimport\u001b[0m \u001b[0mjson\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 6\u001b[0;31m \u001b[0;32mimport\u001b[0m \u001b[0mpypdb\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      7\u001b[0m \u001b[0;32mimport\u001b[0m \u001b[0mtime\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      8\u001b[0m \u001b[0;32mfrom\u001b[0m \u001b[0mprody\u001b[0m \u001b[0;32mimport\u001b[0m \u001b[0;34m*\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mModuleNotFoundError\u001b[0m: No module named 'pypdb'"
     ]
    }
   ],
   "source": [
    "import os\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import glob\n",
    "import json\n",
    "import pypdb\n",
    "import time\n",
    "from prody import *\n",
    "path_to_dud_target_files = os.path.join(*'../../../Benchmarks/dud_targets'.split('/'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio.Seq import Seq\n",
    "from Bio.Blast import NCBIXML\n",
    "from Bio.Blast.Applications import NcbiblastpCommandline\n",
    "\n",
    "def get_uniprot_id_from_pdb_structure(prot_name, verbose = True,\n",
    "                                      path_to_files = F'{path_to_dud_target_files}/'):\n",
    "    # Read the .pdb file and obtains its sequence\n",
    "    prot_name = prot_name.lower()\n",
    "    rec_pdb_file = glob.glob(F'{path_to_files}/{prot_name}/*pdb')[0]\n",
    "    rec_pdb = parsePDB(rec_pdb_file)\n",
    "    resnums = rec_pdb.select('ca').getResnums()\n",
    "    resids = rec_pdb.select('ca').getSequence()\n",
    "    positions = dict(zip(resnums, resids))\n",
    "    sequence = ''\n",
    "    \n",
    "    for i in range(1, resnums[-1]):\n",
    "        try:\n",
    "            sequence += positions[i]\n",
    "        except:\n",
    "            sequence += '-'\n",
    "    while sequence[:5] == '-----':\n",
    "        sequence = sequence[5:]\n",
    "\n",
    "    # Saves the sequence  in a temporal fasta file\n",
    "    with open(\"./temp_fasta.fasta\", \"w\") as ofile:\n",
    "        ofile.write(\">\" + 'Seq' + \"\\n\" + sequence + \"\\n\")\n",
    "        ofile.close()\n",
    "    # Perfroms the blastp over uniprot swissprot\n",
    "    blastx_cline = NcbiblastpCommandline(cmd='blastp', query='temp_fasta.fasta', \n",
    "                    db = os.path.join(*'../../../Benchmarks/uniprot_db/uniprot_sprot.fasta'.split('/')), \n",
    "                    evalue = 0.0001, outfmt = 5, out=\"blast_results.xml\")\n",
    "    blastx_cline()\n",
    "    # Reads the temporary results\n",
    "    blast_record = NCBIXML.read(open(\"blast_results.xml\"))\n",
    "    # Gets the first record, with the highest score and from HUMAN\n",
    "    try:\n",
    "        human_titles = [i.title for i in blast_record.alignments if 'human' in i.title.lower()]\n",
    "        title = human_titles[0]\n",
    "        uniprot_id = title.split('|')[3]\n",
    "        name = title.split('|')[4]\n",
    "    except:\n",
    "        title = blast_record.alignments[0].title\n",
    "        uniprot_id = blast_record.alignments[0].title.split('|')[3]\n",
    "        name = title.split('|')[4] + '*'\n",
    "    if verbose:\n",
    "        print(prot_name, uniprot_id, '\\n', title)\n",
    "    return {'name': name, 'accession': uniprot_id}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 108,
   "metadata": {},
   "outputs": [],
   "source": [
    "# First we have to obtain the UNIPROT Accession numbers\n",
    "# dud_targets_uniprot = {}\n",
    "# for name in dud_targets_names.keys():\n",
    "#     dud_targets_uniprot[name] = get_uniprot_id_from_pdb_structure(prot_name = name)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Saving the result"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# with open('../../../Benchmarks/dud_targets_uniprot_accessions.json', 'w') as fp:\n",
    "#     json.dump(dud_targets_uniprot, fp)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Targets from DEKOIS2.0"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 90,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 115,
   "metadata": {},
   "outputs": [],
   "source": [
    "# dekois_targets_uniprot = {}\n",
    "# for key, pdb_id in dekois_targets_names.items():\n",
    "#     dict_info = pypdb.get_all_info(pdb_id)['polymer']\n",
    "    \n",
    "#     if type(dict_info) is list:\n",
    "#         molecule = dict_info[0]['macroMolecule']\n",
    "#     elif type(dict_info) is dict:\n",
    "#         molecule = dict_info['macroMolecule']\n",
    "#     if type(molecule) is list:\n",
    "#         molecule = molecule[0]\n",
    "#     name = molecule['@name']\n",
    "#     uniprot_id = molecule['accession']['@id']\n",
    "#     dekois_targets_uniprot[key] = {'name': name, 'accession': uniprot_id}\n",
    "#     time.sleep(0.5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 107,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Saving the results\n",
    "# with open('../../../Benchmarks/dekois2_targets_uniprot_accessions.json', 'w') as fp:\n",
    "#     json.dump(dekois_targets_uniprot, fp)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Getting the number of crystals in pdb"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# opening the results\n",
    "dekois_accessions = pd.read_json('../../../Benchmarks/dekois2_targets_uniprot_accessions.json').T\n",
    "dud_accessions = pd.read_json('../../../Benchmarks/dud_targets_uniprot_accessions.json').T"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of targets in DUD set: 21\n",
      "Number of targets in DEKOIS2.0 set: 81\n"
     ]
    }
   ],
   "source": [
    "# Number of targets on each set\n",
    "print(F'Number of targets in DUD set: {dud_accessions.shape[0]}')\n",
    "\n",
    "print(F'Number of targets in DEKOIS2.0 set: {dekois_accessions.shape[0]}')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Then we have to merge the dataframes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Reset the index\n",
    "dud_accessions.reset_index(inplace=True)\n",
    "dekois_accessions.reset_index(inplace=True)\n",
    "\n",
    "# Setting the uniprot accession index as the new index\n",
    "dud_accessions.set_index('accession', inplace = True)\n",
    "dekois_accessions.set_index('accession', inplace = True)\n",
    "\n",
    "# Merging the datasets\n",
    "df_targets = pd.concat([dud_accessions, dekois_accessions], axis = 0).sort_index()\n",
    "df_targets = df_targets.iloc[~df_targets.index.duplicated(keep='last')]\n",
    "# Naming the table\n",
    "df_targets.columns = ['name', 'long_name']\n",
    "\n",
    "# Fill a new columns indicating if the target is in a given datset\n",
    "# for dud\n",
    "df_targets.loc[:, 'DUD'] = [1 if i in dud_accessions.index else 0 for i in df_targets.index]\n",
    "# for dekois\n",
    "df_targets.loc[:, 'DEKOIS2'] = [1 if i in dekois_accessions.index else 0 for i in df_targets.index]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>name</th>\n",
       "      <th>long_name</th>\n",
       "      <th>DUD</th>\n",
       "      <th>DEKOIS2</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>accession</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>O14965</th>\n",
       "      <td>AURKA</td>\n",
       "      <td>Aurora kinase A</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>O15530</th>\n",
       "      <td>PDK1</td>\n",
       "      <td>3-phosphoinositide-dependent protein kinase 1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>O76074</th>\n",
       "      <td>PDE5</td>\n",
       "      <td>cGMP-specific 3',5'-cyclic phosphodiesterase</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P00374</th>\n",
       "      <td>DHFR</td>\n",
       "      <td>Dihydrofolate reductase</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P00533</th>\n",
       "      <td>EGFR</td>\n",
       "      <td>Epidermal growth factor receptor</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Q92731</th>\n",
       "      <td>ER-beta</td>\n",
       "      <td>Estrogen receptor beta</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Q92769</th>\n",
       "      <td>HDAC2</td>\n",
       "      <td>Histone deacetylase 2</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Q9BY41</th>\n",
       "      <td>HDAC8</td>\n",
       "      <td>Histone deacetylase 8</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Q9BYF1</th>\n",
       "      <td>ACE-2</td>\n",
       "      <td>Angiotensin-converting enzyme 2</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Q9P1W9</th>\n",
       "      <td>PIM-2</td>\n",
       "      <td>Serine/threonine-protein kinase pim-2</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>89 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "              name                                      long_name  DUD  \\\n",
       "accession                                                                \n",
       "O14965       AURKA                                Aurora kinase A    0   \n",
       "O15530        PDK1  3-phosphoinositide-dependent protein kinase 1    0   \n",
       "O76074        PDE5   cGMP-specific 3',5'-cyclic phosphodiesterase    0   \n",
       "P00374        DHFR                        Dihydrofolate reductase    1   \n",
       "P00533        EGFR               Epidermal growth factor receptor    1   \n",
       "...            ...                                            ...  ...   \n",
       "Q92731     ER-beta                         Estrogen receptor beta    0   \n",
       "Q92769       HDAC2                          Histone deacetylase 2    0   \n",
       "Q9BY41       HDAC8                          Histone deacetylase 8    0   \n",
       "Q9BYF1       ACE-2                Angiotensin-converting enzyme 2    0   \n",
       "Q9P1W9       PIM-2          Serine/threonine-protein kinase pim-2    0   \n",
       "\n",
       "           DEKOIS2  \n",
       "accession           \n",
       "O14965           1  \n",
       "O15530           1  \n",
       "O76074           1  \n",
       "P00374           1  \n",
       "P00533           1  \n",
       "...            ...  \n",
       "Q92731           1  \n",
       "Q92769           1  \n",
       "Q9BY41           1  \n",
       "Q9BYF1           1  \n",
       "Q9P1W9           1  \n",
       "\n",
       "[89 rows x 4 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_targets"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# df_targets.to_json('targets_table.json')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Get the number of crystals available for each target"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 197,
   "metadata": {},
   "outputs": [],
   "source": [
    "from bs4 import BeautifulSoup\n",
    "import requests\n",
    "\n",
    "def pdb_ids_from_uniprot(uniprot_id, time_sleep = 2):\n",
    "    r = requests.get('https://www.uniprot.org/uniprot/' + uniprot_id)\n",
    "    soup = BeautifulSoup(r.content)\n",
    "    pdb_tags = soup.find_all('a', {'class': 'pdb'})\n",
    "    \n",
    "    pdb_ids = [tag.get_text() for tag in pdb_tags]\n",
    "    \n",
    "    result_dict = {'pdb_ids': pdb_ids, 'n_crystals': len(pdb_ids)}\n",
    "    \n",
    "    print(uniprot_id, len(pdb_ids))\n",
    "    time.sleep(time_sleep)\n",
    "    return(result_dict)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 198,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "O14965 155\n",
      "O15530 68\n",
      "O76074 37\n",
      "P00374 79\n",
      "P00533 189\n",
      "P00720 702\n",
      "P00734 391\n",
      "P00742 144\n",
      "P00749 143\n",
      "P00750 11\n",
      "P00811 102\n",
      "P00813 2\n",
      "P03366 323\n",
      "P03372 278\n",
      "P04035 22\n",
      "P04058 117\n",
      "P04062 28\n",
      "P04150 43\n",
      "P04183 4\n",
      "P04585 204\n",
      "P04587 71\n",
      "P04626 34\n",
      "P04818 59\n",
      "P05979 30\n",
      "P06239 55\n",
      "P06401 20\n",
      "P06737 18\n",
      "P07711 36\n",
      "P07900 295\n",
      "P08069 25\n",
      "P08253 11\n",
      "P09874 62\n",
      "P0C6X7 100\n",
      "P10275 82\n",
      "P10415 26\n",
      "P11217 1\n",
      "P11309 156\n",
      "P11362 65\n",
      "P11509 11\n",
      "P12821 44\n",
      "P12931 64\n",
      "P14061 27\n",
      "P15056 74\n",
      "P15121 145\n",
      "P17948 12\n",
      "P19793 87\n",
      "P19971 4\n",
      "P21964 12\n",
      "P22102 29\n",
      "P22303 42\n",
      "P23219 0\n",
      "P24941 410\n",
      "P27907 5\n",
      "P28845 40\n",
      "P31749 30\n",
      "P35354 8\n",
      "P35968 52\n",
      "P43235 56\n",
      "P45983 32\n",
      "P45984 2\n",
      "P48736 93\n",
      "P49137 25\n",
      "P49841 84\n",
      "P52333 34\n",
      "P52732 55\n",
      "P53779 51\n",
      "P54760 23\n",
      "P55859 21\n",
      "P62942 50\n",
      "P78536 23\n",
      "P80276 5\n",
      "P9WGR1 90\n",
      "Q00987 103\n",
      "Q02763 15\n",
      "Q04759 8\n",
      "Q05769 49\n",
      "Q07343 43\n",
      "Q07869 18\n",
      "Q08881 36\n",
      "Q13464 27\n",
      "Q16539 236\n",
      "Q16769 24\n",
      "Q6DE08 12\n",
      "Q8IXJ6 29\n",
      "Q92731 32\n",
      "Q92769 6\n",
      "Q9BY41 47\n",
      "Q9BYF1 16\n",
      "Q9P1W9 2\n"
     ]
    }
   ],
   "source": [
    "targets_pdb_ids = {}\n",
    "for uniprot_id in df_targets.index:\n",
    "    targets_pdb_ids[uniprot_id] = pdb_ids_from_uniprot(uniprot_id)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 203,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>pdb_ids</th>\n",
       "      <th>n_crystals</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>O14965</th>\n",
       "      <td>[1MQ4, 1MUO, 1OL5, 1OL6, 1OL7, 2BMC, 2C6D, 2C6...</td>\n",
       "      <td>155</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>O15530</th>\n",
       "      <td>[1H1W, 1OKY, 1OKZ, 1UU3, 1UU7, 1UU8, 1UU9, 1UV...</td>\n",
       "      <td>68</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>O76074</th>\n",
       "      <td>[1RKP, 1T9R, 1T9S, 1TBF, 1UDT, 1UDU, 1UHO, 1XO...</td>\n",
       "      <td>37</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P00374</th>\n",
       "      <td>[1BOZ, 1DHF, 1DLR, 1DLS, 1DRF, 1HFP, 1HFQ, 1HF...</td>\n",
       "      <td>79</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P00533</th>\n",
       "      <td>[1DNQ, 1DNR, 1IVO, 1M14, 1M17, 1MOX, 1NQL, 1XK...</td>\n",
       "      <td>189</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Q92731</th>\n",
       "      <td>[1L2J, 1NDE, 1QKM, 1U3Q, 1U3R, 1U3S, 1U9E, 1X7...</td>\n",
       "      <td>32</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Q92769</th>\n",
       "      <td>[3MAX, 4LXZ, 4LY1, 5IWG, 5IX0, 6G3O]</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Q9BY41</th>\n",
       "      <td>[1T64, 1T67, 1T69, 1VKG, 1W22, 2V5W, 2V5X, 3EW...</td>\n",
       "      <td>47</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Q9BYF1</th>\n",
       "      <td>[1R42, 1R4L, 1XJP, 2AJF, 3D0G, 3D0H, 3D0I, 3KB...</td>\n",
       "      <td>16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Q9P1W9</th>\n",
       "      <td>[2IWI, 4X7Q]</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>89 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                  pdb_ids n_crystals\n",
       "O14965  [1MQ4, 1MUO, 1OL5, 1OL6, 1OL7, 2BMC, 2C6D, 2C6...        155\n",
       "O15530  [1H1W, 1OKY, 1OKZ, 1UU3, 1UU7, 1UU8, 1UU9, 1UV...         68\n",
       "O76074  [1RKP, 1T9R, 1T9S, 1TBF, 1UDT, 1UDU, 1UHO, 1XO...         37\n",
       "P00374  [1BOZ, 1DHF, 1DLR, 1DLS, 1DRF, 1HFP, 1HFQ, 1HF...         79\n",
       "P00533  [1DNQ, 1DNR, 1IVO, 1M14, 1M17, 1MOX, 1NQL, 1XK...        189\n",
       "...                                                   ...        ...\n",
       "Q92731  [1L2J, 1NDE, 1QKM, 1U3Q, 1U3R, 1U3S, 1U9E, 1X7...         32\n",
       "Q92769               [3MAX, 4LXZ, 4LY1, 5IWG, 5IX0, 6G3O]          6\n",
       "Q9BY41  [1T64, 1T67, 1T69, 1VKG, 1W22, 2V5W, 2V5X, 3EW...         47\n",
       "Q9BYF1  [1R42, 1R4L, 1XJP, 2AJF, 3D0G, 3D0H, 3D0I, 3KB...         16\n",
       "Q9P1W9                                       [2IWI, 4X7Q]          2\n",
       "\n",
       "[89 rows x 2 columns]"
      ]
     },
     "execution_count": 203,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pd.DataFrame(targets_pdb_ids).T.to_json('targets_PDB_IDS.json')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Old functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 417,
   "metadata": {},
   "outputs": [],
   "source": [
    "# import requests\n",
    "# from Bio import SeqIO\n",
    "# import pickle\n",
    "# import pypdb\n",
    "\n",
    "def _blast_pdb(uniprot_id, identidad = 95):\n",
    "    \n",
    "    file_blast_record = 'target_fastas/' + uniprot_id + '_blast_record.pkl'\n",
    "    if os.path.isfile(file_blast_record):\n",
    "        blast_record = pickle.load(open(file_blast_record, 'rb'))\n",
    "    else:\n",
    "        # Descargamos el fasta de UniProt\n",
    "        url_fasta = requests.get(\"https://www.uniprot.org/uniprot/\" + uniprot_id + \".fasta\")\n",
    "        file_name_fasta = 'target_fastas/' + uniprot_id + '.fasta'\n",
    "        open(file_name_fasta, 'wb').write(url_fasta.content)\n",
    "        # Leemos la secuenciade aminoácidos\n",
    "        fasta_prot = SeqIO.read(open(file_name_fasta),'fasta')\n",
    "        seq_prot = str(fasta_prot.seq)\n",
    "        try:\n",
    "            blast_record = blastPDB(seq_prot)\n",
    "            pickle.dump(blast_record, open(file_blast_record, 'wb'))\n",
    "        except: pass\n",
    "    \n",
    "    pdbids = blast_record.getHits(percent_identity = identidad) # Devuelve un diccionario con cada proteína hit\n",
    "    len(pdbids) # num de proteínas que tienen un 95% de identidad con la secuencia\n",
    "    pdbids_list = list(pdbids.keys()) # lista de pdb_ids de las estructuras de la proteina\n",
    "    print(F'{len(pdbids_list)} hits con una identidad del {identidad}%')\n",
    "    time.sleep(2)\n",
    "    return(pdbids_list)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "metadata": {},
   "outputs": [],
   "source": [
    "# import requests\n",
    "# from Bio import SeqIO\n",
    "# import pickle\n",
    "# import pypdb\n",
    "\n",
    "def _blast_pdb(uniprot_id, identidad = 95):\n",
    "    \n",
    "    file_blast_record = 'target_fastas/' + uniprot_id + '_blast_record.pkl'\n",
    "    if os.path.isfile(file_blast_record):\n",
    "        blast_record = pickle.load(open(file_blast_record, 'rb'))\n",
    "    else:\n",
    "        # Descargamos el fasta de UniProt\n",
    "        url_fasta = requests.get(\"https://www.uniprot.org/uniprot/\" + uniprot_id + \".fasta\")\n",
    "        fasta = ''.join(url_fasta.text.split('\\n')[1:])\n",
    "        seq_prot = fasta\n",
    "        try:\n",
    "            blast_record = blastPDB(seq_prot)\n",
    "            pickle.dump(blast_record, open(file_blast_record, 'wb'))\n",
    "        except: pass\n",
    "    \n",
    "    pdbids = blast_record.getHits(percent_identity = identidad) # Devuelve un diccionario con cada proteína hit\n",
    "    len(pdbids) # num de proteínas que tienen un 95% de identidad con la secuencia\n",
    "    pdbids_list = list(pdbids.keys()) # lista de pdb_ids de las estructuras de la proteina\n",
    "    print(F'{len(pdbids_list)} hits con una identidad del {identidad}%')\n",
    "    time.sleep(2)\n",
    "    return(pdbids_list)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 174,
   "metadata": {},
   "outputs": [],
   "source": [
    "# import pypdb\n",
    "# import untangle\n",
    "# import requests\n",
    "# import pandas as pd\n",
    "# import numpy as np\n",
    "\n",
    "def _blast_pdb(uniprot_id, eCutoff = 0, time_sleep = 2):\n",
    "    url_fasta = requests.get(\"https://www.uniprot.org/uniprot/\" + uniprot_id + \".fasta\")\n",
    "    seq = ''.join(url_fasta.text.split('\\n')[1:])\n",
    "\n",
    "    payload = {\n",
    "                'sequence': seq,\n",
    "                'eCutoff':str(eCutoff),\n",
    "                'matrix':'BLOSUM62',\n",
    "                'outputFormat':'XML' # or 'HTML' or nothing for text format\n",
    "                }\n",
    "\n",
    "    pdbUrl = 'http://www.rcsb.org/pdb/rest/getBlastPDB1'\n",
    "\n",
    "    # Create the url with parameters\n",
    "    r = requests.get(pdbUrl, params = payload)\n",
    "    # Parse \n",
    "    soup = BeautifulSoup(r.content)\n",
    "\n",
    "    pdb_ids = []\n",
    "    for hit in soup.find_all('hit'):\n",
    "        pdb_id = hit.find('hit_def').get_text().split(':')[0]\n",
    "        pdb_ids.append(pdb_id)\n",
    "    #     e_val = float(hit.find('hsp_evalue').get_text())\n",
    "    #     hit_len = int(hit.find('hit_len').get_text())\n",
    "    print(uniprot_id, len(pdb_ids))\n",
    "    \n",
    "    time.sleep(time_sleep)\n",
    "    return({'pdb_ids': pdb_ids, 'n_crystals': len(pdb_ids)})\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "hideCode": false,
    "hideOutput": true,
    "hidePrompt": false
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {
    "hideCode": false,
    "hidePrompt": false
   },
   "outputs": [],
   "source": [
    "dud_targets_names = {\n",
    "    \"ACE\": \"Angiotensin-converting enzyme\",\n",
    "    \"ACHE\": \"Acetylcholine esterase\",\n",
    "    \"ADA\": \"Adenosine deaminase\",\n",
    "    \"ALR2\": \"Aldose reductase\",\n",
    "    \"AmpC\": \"AmpC beta lactamase\",\n",
    "    \"AR\": \"Androgen receptor\",\n",
    "    \"CDK2\": \"Cyclin dependent kinase 2\",\n",
    "    \"COMT\": \"Catechol O-methyltransferase\",\n",
    "    \"COX1\": \"Cyclooxygenase 1\",\n",
    "    \"COX2\": \"Cyclooxygenase 2\",\n",
    "    \"DHFR\": \"Dihydrofolate reductase\",\n",
    "    \"EGFr\": \"Epidermal growth factor receptor kinase\",\n",
    "    \"ER_agonist\": \"Estrogen receptor agonist\",\n",
    "    \"ER_antagonist\": \"Estrogen receptor antagonist\",\n",
    "    \"FGFr1\": \"Fibroblast growth factor receptor kinase\",\n",
    "    \"FXa\": \"Factor Xa\",\n",
    "    \"GART\": \"glycinamide ribonucleotide transformylase\",\n",
    "    \"GPB\": \"Glycogen phosphorylase beta\",\n",
    "    \"GR\": \"Glutocorticoid receptor\",\n",
    "    \"HIVPR\": \"HIV protease\",\n",
    "    \"HIVRT\": \"HIV reverse transcriptase\",\n",
    "    \"HMGR\": \"Hydroxymethylglutaryl-CoA reductase\",\n",
    "    \"HSP90\": \"Human heat shock protein 90 kinase\",\n",
    "    \"InhA\": \"Enoyl ACP reductase\",\n",
    "    \"MR\": \"Mineralcorticoid receptor\",\n",
    "    \"NA\": \"Neuraminidase\",\n",
    "    \"P38\": \"P38 mitogen activated protein kinase\",\n",
    "    \"PARP\": \"Poly(ADP-ribose) polymerase\",\n",
    "    \"PDE5\": \"Phosphodiesterase V\",\n",
    "    \"PDGFrb\": \"Platlet derived growth factor receptor kinase\",\n",
    "    \"PNP\": \"Purine nucleoside phosphorylase\",\n",
    "    \"PPARg\": \"Peroxisome proliferator activated receptor gamma\",\n",
    "    \"PR\": \"Progesterone receptor\",\n",
    "    \"RXRa\": \"Retinoic X receptor alpha\",\n",
    "    \"SAHH\": \"S-adenosyl-homocysteine hydrolase\",\n",
    "    \"SRC\": \"Tyrosine kinase SRC\",\n",
    "    \"Thrombin\": \"Thrombin\",\n",
    "    \"TK\": \"Thymidine kinase\",\n",
    "    \"Trypsin\": \"Trypsin\",\n",
    "    \"VEGFr2\": \"Vascular endothelial growth factor receptor kinase\"\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "dekois_targets_names = {\n",
    "    \"11-beta-HSD1\": \"3tfq\",\n",
    "    \"17-beta-HSD1\": \"3klm\",\n",
    "    \"A2A\": \"3eml\",\n",
    "    \"ACE\": \"1uze\",\n",
    "    \"ACE-2\": \"1r4l\",\n",
    "    \"ACHE\": \"1eve\",\n",
    "    \"ADAM17\": \"3ewj\",\n",
    "    \"ADRB2\": \"3ny9\",\n",
    "    \"AKT1\": \"3qkl\",\n",
    "    \"ALR2\": \"1ah3\",\n",
    "    \"AR\": \"1e3g\",\n",
    "    \"AURKA\": \"3fdn\",\n",
    "    \"AURKB\": \"2vgo\",\n",
    "    \"Bcl-2\": \"2w3l\",\n",
    "    \"BRAF\": \"3skc\",\n",
    "    \"CATL\": \"3bc3\",\n",
    "    \"CDK2\": \"1ckp\",\n",
    "    \"COX1\": \"3kk6\",\n",
    "    \"COX2\": \"1cx2\",\n",
    "    \"CTSK\": \"3kx1\",\n",
    "    \"CYP2A6\": \"1z11\",\n",
    "    \"DHFR\": \"1s3v\",\n",
    "    \"EGFR\": \"1m17\",\n",
    "    \"EPHB4\": \"2vwz\",\n",
    "    \"ERBB2\": \"3pp0\",\n",
    "    \"ER-beta\": \"3oll\",\n",
    "    \"FGFR1\": \"1agw\",\n",
    "    \"FKBP1A\": \"2dg3\",\n",
    "    \"FXA\": \"1f0r\",\n",
    "    \"GBA\": \"2wcg\",\n",
    "    \"GR\": \"1nhz\",\n",
    "    \"GSK3B\": \"3i4b\",\n",
    "    \"HDAC2\": \"3max\",\n",
    "    \"HDAC8\": \"3sff\",\n",
    "    \"HIV1PR\": \"3nu3\",\n",
    "    \"HIV1RT\": \"1s6p\",\n",
    "    \"HMGA\": \"1hw8\",\n",
    "    \"HSP90\": \"1uy6\",\n",
    "    \"IGF1R\": \"3nw7\",\n",
    "    \"INHA\": \"1p44\",\n",
    "    \"ITK\": \"3mj1\",\n",
    "    \"JAK3\": \"3lxl\",\n",
    "    \"JNK1\": \"3elj\",\n",
    "    \"JNK2\": \"3npc\",\n",
    "    \"JNK3\": \"2b1p\",\n",
    "    \"KIF11\": \"3k5e\",\n",
    "    \"Lck\": \"3mpm\",\n",
    "    \"MDM2\": \"3lbk\",\n",
    "    \"MK2\": \"3kc3\",\n",
    "    \"MMP2\": \"1hov\",\n",
    "    \"NA\": \"1a4g\",\n",
    "    \"P38\": \"1ouk\",\n",
    "    \"PARP-1\": \"3l3m\",\n",
    "    \"PDE4B\": \"3frg\",\n",
    "    \"PDE5\": \"1xp0\",\n",
    "    \"PDK1\": \"2xch\",\n",
    "    \"PI3Kg\": \"3dbs\",\n",
    "    \"PIM-1\": \"3r04\",\n",
    "    \"PIM-2\": \"2iwi\",\n",
    "    \"PNP\": \"1b8o\",\n",
    "    \"PPARa\": \"2p54\",\n",
    "    \"PPARg\": \"1fm9\",\n",
    "    \"PR\": \"2w8y\",\n",
    "    \"PRKCQ\": \"1xjd\",\n",
    "    \"PYGL-in\": \"1xoi\",\n",
    "    \"PYGL-out\": \"3dds\",\n",
    "    \"QPCT\": \"2afx\",\n",
    "    \"ROCK-1\": \"3v8s\",\n",
    "    \"RXR\": \"2p1t\",\n",
    "    \"SARS-HCoV\": \"2z94\",\n",
    "    \"SIRT2\": \"1j8f\",\n",
    "    \"SRC\": \"2src\",\n",
    "    \"Thrombin\": \"3rm2\",\n",
    "    \"TIE2\": \"2oo8\",\n",
    "    \"TK\": \"1w4r\",\n",
    "    \"TP\": \"1uou\",\n",
    "    \"TPA\": \"1a5h\",\n",
    "    \"TS\": \"1i00\",\n",
    "    \"uPA\": \"3mhw\",\n",
    "    \"VEGFR1\": \"3hng\",\n",
    "    \"VEGFR2\": \"3c7q\"\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "81"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(dekois_targets_names)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "hide_code_all_hidden": false,
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
