{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Protein targets"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "hideCode": false,
    "hidePrompt": false
   },
   "outputs": [],
   "source": [
    "# Cargar cada estructura desde el archivo que proporciona DUD"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import glob\n",
    "import json\n",
    "import pypdb\n",
    "import time\n",
    "from prody import *\n",
    "path_to_dud_target_files = os.path.join(*'../../../Benchmarks/dud_targets'.split('/'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio.Seq import Seq\n",
    "from Bio.Blast import NCBIXML\n",
    "from Bio.Blast.Applications import NcbiblastpCommandline\n",
    "\n",
    "def get_uniprot_id_from_pdb_structure(prot_name, verbose = True,\n",
    "                                      path_to_files = F'{path_to_dud_target_files}/'):\n",
    "    # Read the .pdb file and obtains its sequence\n",
    "    prot_name = prot_name.lower()\n",
    "    rec_pdb_file = glob.glob(F'{path_to_files}/{prot_name}/*pdb')[0]\n",
    "    rec_pdb = parsePDB(rec_pdb_file)\n",
    "    resnums = rec_pdb.select('ca').getResnums()\n",
    "    resids = rec_pdb.select('ca').getSequence()\n",
    "    positions = dict(zip(resnums, resids))\n",
    "    sequence = ''\n",
    "    \n",
    "    for i in range(1, resnums[-1]):\n",
    "        try:\n",
    "            sequence += positions[i]\n",
    "        except:\n",
    "            sequence += '-'\n",
    "    while sequence[:5] == '-----':\n",
    "        sequence = sequence[5:]\n",
    "\n",
    "    # Saves the sequence  in a temporal fasta file\n",
    "    with open(\"./temp_fasta.fasta\", \"w\") as ofile:\n",
    "        ofile.write(\">\" + 'Seq' + \"\\n\" + sequence + \"\\n\")\n",
    "        ofile.close()\n",
    "    # Perfroms the blastp over uniprot swissprot\n",
    "    blastx_cline = NcbiblastpCommandline(cmd='blastp', query='temp_fasta.fasta', \n",
    "                    db = os.path.join(*'../../../Benchmarks/uniprot_db/uniprot_sprot.fasta'.split('/')), \n",
    "                    evalue = 0.0001, outfmt = 5, out=\"blast_results.xml\")\n",
    "    blastx_cline()\n",
    "    # Reads the temporary results\n",
    "    blast_record = NCBIXML.read(open(\"blast_results.xml\"))\n",
    "    # Gets the first record, with the highest score and from HUMAN\n",
    "    try:\n",
    "        human_titles = [i.title for i in blast_record.alignments if 'human' in i.title.lower()]\n",
    "        title = human_titles[0]\n",
    "        uniprot_id = title.split('|')[3]\n",
    "        name = title.split('|')[4]\n",
    "    except:\n",
    "        title = blast_record.alignments[0].title\n",
    "        uniprot_id = blast_record.alignments[0].title.split('|')[3]\n",
    "        name = title.split('|')[4] + '*'\n",
    "    if verbose:\n",
    "        print(prot_name, uniprot_id, '\\n', title)\n",
    "    return {'name': name, 'accession': uniprot_id}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "# First we have to obtain the UNIPROT Accession numbers\n",
    "# dud_targets_uniprot = {}\n",
    "# for name in dud_targets_names.keys():\n",
    "#     dud_targets_uniprot[name] = get_uniprot_id_from_pdb_structure(prot_name = name)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Saving the result"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# with open('../../../Benchmarks/dud_targets_uniprot_accessions.json', 'w') as fp:\n",
    "#     json.dump(dud_targets_uniprot, fp)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Targets from DEKOIS2.0"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "dekois_targets_uniprot = {}\n",
    "for key, pdb_id in dekois_targets_names.items():\n",
    "    dict_info = pypdb.get_all_info(pdb_id)['polymer']\n",
    "    \n",
    "    if type(dict_info) is list:\n",
    "        molecule = dict_info[0]['macroMolecule']\n",
    "    elif type(dict_info) is dict:\n",
    "        molecule = dict_info['macroMolecule']\n",
    "    if type(molecule) is list:\n",
    "        molecule = molecule[0]\n",
    "    name = molecule['@name']\n",
    "    uniprot_id = molecule['accession']['@id']\n",
    "    dekois_targets_uniprot[key] = {'name': name, 'accession': uniprot_id}\n",
    "    time.sleep(0.5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'11-beta-HSD1': {'name': 'Corticosteroid 11-beta-dehydrogenase isozyme 1',\n",
       "  'accession': 'P28845'},\n",
       " '17-beta-HSD1': {'name': 'Estradiol 17-beta-dehydrogenase 1',\n",
       "  'accession': 'P14061'},\n",
       " 'A2A': {'name': 'Endolysin', 'accession': 'P00720'},\n",
       " 'ACE': {'name': 'Angiotensin-converting enzyme', 'accession': 'P12821'},\n",
       " 'ACE-2': {'name': 'Angiotensin-converting enzyme 2', 'accession': 'Q9BYF1'},\n",
       " 'ACHE': {'name': 'Acetylcholinesterase', 'accession': 'P04058'},\n",
       " 'ADAM17': {'name': 'Disintegrin and metalloproteinase domain-containing protein 17',\n",
       "  'accession': 'P78536'},\n",
       " 'ADRB2': {'name': 'Endolysin', 'accession': 'P00720'},\n",
       " 'AKT1': {'name': 'RAC-alpha serine/threonine-protein kinase',\n",
       "  'accession': 'P31749'},\n",
       " 'ALR2': {'name': 'Aldose reductase', 'accession': 'P80276'},\n",
       " 'AR': {'name': 'Androgen receptor', 'accession': 'P10275'},\n",
       " 'AURKA': {'name': 'Aurora kinase A', 'accession': 'O14965'},\n",
       " 'AURKB': {'name': 'Aurora kinase B-A', 'accession': 'Q6DE08'},\n",
       " 'Bcl-2': {'name': 'Apoptosis regulator Bcl-2', 'accession': 'P10415'},\n",
       " 'BRAF': {'name': 'Serine/threonine-protein kinase B-raf',\n",
       "  'accession': 'P15056'},\n",
       " 'CATL': {'name': 'Cathepsin L1', 'accession': 'P07711'},\n",
       " 'CDK2': {'name': 'Cyclin-dependent kinase 2', 'accession': 'P24941'},\n",
       " 'COX1': {'name': 'Prostaglandin G/H synthase 1', 'accession': 'P05979'},\n",
       " 'COX2': {'name': 'Prostaglandin G/H synthase 2', 'accession': 'Q05769'},\n",
       " 'CTSK': {'name': 'Cathepsin K', 'accession': 'P43235'},\n",
       " 'CYP2A6': {'name': 'Cytochrome P450 2A6', 'accession': 'P11509'},\n",
       " 'DHFR': {'name': 'Dihydrofolate reductase', 'accession': 'P00374'},\n",
       " 'EGFR': {'name': 'Epidermal growth factor receptor', 'accession': 'P00533'},\n",
       " 'EPHB4': {'name': 'Ephrin type-B receptor 4', 'accession': 'P54760'},\n",
       " 'ERBB2': {'name': 'Receptor tyrosine-protein kinase erbB-2',\n",
       "  'accession': 'P04626'},\n",
       " 'ER-beta': {'name': 'Estrogen receptor beta', 'accession': 'Q92731'},\n",
       " 'FGFR1': {'name': 'Fibroblast growth factor receptor 1',\n",
       "  'accession': 'P11362'},\n",
       " 'FKBP1A': {'name': 'Peptidyl-prolyl cis-trans isomerase FKBP1A',\n",
       "  'accession': 'P62942'},\n",
       " 'FXA': {'name': 'Coagulation factor X', 'accession': 'P00742'},\n",
       " 'GBA': {'name': 'Glucosylceramidase', 'accession': 'P04062'},\n",
       " 'GR': {'name': 'Glucocorticoid receptor', 'accession': 'P04150'},\n",
       " 'GSK3B': {'name': 'Glycogen synthase kinase-3 beta', 'accession': 'P49841'},\n",
       " 'HDAC2': {'name': 'Histone deacetylase 2', 'accession': 'Q92769'},\n",
       " 'HDAC8': {'name': 'Histone deacetylase 8', 'accession': 'Q9BY41'},\n",
       " 'HIV1PR': {'name': 'Gag-Pol polyprotein', 'accession': 'P03366'},\n",
       " 'HIV1RT': {'name': 'Gag-Pol polyprotein', 'accession': 'P03366'},\n",
       " 'HMGA': {'name': '3-hydroxy-3-methylglutaryl-coenzyme A reductase',\n",
       "  'accession': 'P04035'},\n",
       " 'HSP90': {'name': 'Heat shock protein HSP 90-alpha', 'accession': 'P07900'},\n",
       " 'IGF1R': {'name': 'Insulin-like growth factor 1 receptor',\n",
       "  'accession': 'P08069'},\n",
       " 'INHA': {'name': 'Enoyl-[acyl-carrier-protein] reductase [NADH]',\n",
       "  'accession': 'P9WGR1'},\n",
       " 'ITK': {'name': 'Tyrosine-protein kinase ITK/TSK', 'accession': 'Q08881'},\n",
       " 'JAK3': {'name': 'Tyrosine-protein kinase JAK3', 'accession': 'P52333'},\n",
       " 'JNK1': {'name': 'Mitogen-activated protein kinase 8', 'accession': 'P45983'},\n",
       " 'JNK2': {'name': 'Mitogen-activated protein kinase 9', 'accession': 'P45984'},\n",
       " 'JNK3': {'name': 'Mitogen-activated protein kinase 10',\n",
       "  'accession': 'P53779'},\n",
       " 'KIF11': {'name': 'Kinesin-like protein KIF11', 'accession': 'P52732'},\n",
       " 'Lck': {'name': 'Tyrosine-protein kinase Lck', 'accession': 'P06239'},\n",
       " 'MDM2': {'name': 'E3 ubiquitin-protein ligase Mdm2', 'accession': 'Q00987'},\n",
       " 'MK2': {'name': 'MAP kinase-activated protein kinase 2',\n",
       "  'accession': 'P49137'},\n",
       " 'MMP2': {'name': '72 kDa type IV collagenase', 'accession': 'P08253'},\n",
       " 'NA': {'name': 'Neuraminidase', 'accession': 'P27907'},\n",
       " 'P38': {'name': 'Mitogen-activated protein kinase 14', 'accession': 'Q16539'},\n",
       " 'PARP-1': {'name': 'Poly [ADP-ribose] polymerase 1', 'accession': 'P09874'},\n",
       " 'PDE4B': {'name': \"cAMP-specific 3',5'-cyclic phosphodiesterase 4B\",\n",
       "  'accession': 'Q07343'},\n",
       " 'PDE5': {'name': \"cGMP-specific 3',5'-cyclic phosphodiesterase\",\n",
       "  'accession': 'O76074'},\n",
       " 'PDK1': {'name': '3-phosphoinositide-dependent protein kinase 1',\n",
       "  'accession': 'O15530'},\n",
       " 'PI3Kg': {'name': 'Phosphatidylinositol 4,5-bisphosphate 3-kinase catalytic subunit gamma isoform',\n",
       "  'accession': 'P48736'},\n",
       " 'PIM-1': {'name': 'Serine/threonine-protein kinase pim-1',\n",
       "  'accession': 'P11309'},\n",
       " 'PIM-2': {'name': 'Serine/threonine-protein kinase pim-2',\n",
       "  'accession': 'Q9P1W9'},\n",
       " 'PNP': {'name': 'Purine nucleoside phosphorylase', 'accession': 'P55859'},\n",
       " 'PPARa': {'name': 'Peroxisome proliferator-activated receptor alpha',\n",
       "  'accession': 'Q07869'},\n",
       " 'PPARg': {'name': 'Retinoic acid receptor RXR-alpha', 'accession': 'P19793'},\n",
       " 'PR': {'name': 'Progesterone receptor', 'accession': 'P06401'},\n",
       " 'PRKCQ': {'name': 'Protein kinase C theta type', 'accession': 'Q04759'},\n",
       " 'PYGL-in': {'name': 'Glycogen phosphorylase, liver form',\n",
       "  'accession': 'P06737'},\n",
       " 'PYGL-out': {'name': 'Glycogen phosphorylase, liver form',\n",
       "  'accession': 'P06737'},\n",
       " 'QPCT': {'name': 'Glutaminyl-peptide cyclotransferase',\n",
       "  'accession': 'Q16769'},\n",
       " 'ROCK-1': {'name': 'Rho-associated protein kinase 1', 'accession': 'Q13464'},\n",
       " 'RXR': {'name': 'Retinoic acid receptor RXR-alpha', 'accession': 'P19793'},\n",
       " 'SARS-HCoV': {'name': 'Replicase polyprotein 1ab', 'accession': 'P0C6X7'},\n",
       " 'SIRT2': {'name': 'NAD-dependent protein deacetylase sirtuin-2',\n",
       "  'accession': 'Q8IXJ6'},\n",
       " 'SRC': {'name': 'Proto-oncogene tyrosine-protein kinase Src',\n",
       "  'accession': 'P12931'},\n",
       " 'Thrombin': {'name': 'Prothrombin', 'accession': 'P00734'},\n",
       " 'TIE2': {'name': 'Angiopoietin-1 receptor', 'accession': 'Q02763'},\n",
       " 'TK': {'name': 'Thymidine kinase, cytosolic', 'accession': 'P04183'},\n",
       " 'TP': {'name': 'Thymidine phosphorylase', 'accession': 'P19971'},\n",
       " 'TPA': {'name': 'Tissue-type plasminogen activator', 'accession': 'P00750'},\n",
       " 'TS': {'name': 'Thymidylate synthase', 'accession': 'P04818'},\n",
       " 'uPA': {'name': 'Urokinase-type plasminogen activator',\n",
       "  'accession': 'P00749'},\n",
       " 'VEGFR1': {'name': 'Vascular endothelial growth factor receptor 1',\n",
       "  'accession': 'P17948'},\n",
       " 'VEGFR2': {'name': 'Vascular endothelial growth factor receptor 2',\n",
       "  'accession': 'P35968'}}"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dekois_targets_uniprot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 107,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Saving the results\n",
    "# with open('../../../Benchmarks/dekois2_targets_uniprot_accessions.json', 'w') as fp:\n",
    "#     json.dump(dekois_targets_uniprot, fp)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Getting the number of crystals in pdb"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "# opening the results\n",
    "dekois_accessions = pd.read_json('../../../Benchmarks/datasets/dekois2_targets_uniprot_accessions.json').T\n",
    "dud_accessions = pd.read_json('../../../Benchmarks/datasets/dud_targets_uniprot_accessions.json').T\n",
    "\n",
    "# Reset the index\n",
    "dud_accessions.reset_index(inplace=True)\n",
    "dekois_accessions.reset_index(inplace=True)\n",
    "\n",
    "# Setting the uniprot accession index as the new index\n",
    "dud_accessions.set_index('accession', inplace = True)\n",
    "dekois_accessions.set_index('accession', inplace = True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of targets in DUD set: 21\n",
      "Number of targets in DEKOIS2.0 set: 81\n"
     ]
    }
   ],
   "source": [
    "# Number of targets on each set\n",
    "print(F'Number of targets in DUD set: {dud_accessions.shape[0]}')\n",
    "\n",
    "print(F'Number of targets in DEKOIS2.0 set: {dekois_accessions.shape[0]}')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Then we have to merge the dataframes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>index</th>\n",
       "      <th>name</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>accession</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>P17948</th>\n",
       "      <td>VEGFR1</td>\n",
       "      <td>Vascular endothelial growth factor receptor 1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P19793</th>\n",
       "      <td>RXR</td>\n",
       "      <td>Retinoic acid receptor RXR-alpha</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P19971</th>\n",
       "      <td>TP</td>\n",
       "      <td>Thymidine phosphorylase</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P24941</th>\n",
       "      <td>CDK2</td>\n",
       "      <td>Cyclin-dependent kinase 2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P27907</th>\n",
       "      <td>NA</td>\n",
       "      <td>Neuraminidase</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P28845</th>\n",
       "      <td>11-beta-HSD1</td>\n",
       "      <td>Corticosteroid 11-beta-dehydrogenase isozyme 1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P29274</th>\n",
       "      <td>A2A</td>\n",
       "      <td>Adenosine receptor A2a</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P31749</th>\n",
       "      <td>AKT1</td>\n",
       "      <td>RAC-alpha serine/threonine-protein kinase</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P35968</th>\n",
       "      <td>VEGFR2</td>\n",
       "      <td>Vascular endothelial growth factor receptor 2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P37231</th>\n",
       "      <td>PPARg</td>\n",
       "      <td>Peroxisome proliferator-activated receptor gamma</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P43235</th>\n",
       "      <td>CTSK</td>\n",
       "      <td>Cathepsin K</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P45983</th>\n",
       "      <td>JNK1</td>\n",
       "      <td>Mitogen-activated protein kinase 8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P45984</th>\n",
       "      <td>JNK2</td>\n",
       "      <td>Mitogen-activated protein kinase 9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P48736</th>\n",
       "      <td>PI3Kg</td>\n",
       "      <td>Phosphatidylinositol 4,5-bisphosphate 3-kinase...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P49137</th>\n",
       "      <td>MK2</td>\n",
       "      <td>MAP kinase-activated protein kinase 2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P49841</th>\n",
       "      <td>GSK3B</td>\n",
       "      <td>Glycogen synthase kinase-3 beta</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P52333</th>\n",
       "      <td>JAK3</td>\n",
       "      <td>Tyrosine-protein kinase JAK3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P52732</th>\n",
       "      <td>KIF11</td>\n",
       "      <td>Kinesin-like protein KIF11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P53779</th>\n",
       "      <td>JNK3</td>\n",
       "      <td>Mitogen-activated protein kinase 10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P54760</th>\n",
       "      <td>EPHB4</td>\n",
       "      <td>Ephrin type-B receptor 4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P55859</th>\n",
       "      <td>PNP</td>\n",
       "      <td>Purine nucleoside phosphorylase</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P62942</th>\n",
       "      <td>FKBP1A</td>\n",
       "      <td>Peptidyl-prolyl cis-trans isomerase FKBP1A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P78536</th>\n",
       "      <td>ADAM17</td>\n",
       "      <td>Disintegrin and metalloproteinase domain-conta...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P80276</th>\n",
       "      <td>ALR2</td>\n",
       "      <td>Aldose reductase</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P9WGR1</th>\n",
       "      <td>INHA</td>\n",
       "      <td>Enoyl-[acyl-carrier-protein] reductase [NADH]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Q00987</th>\n",
       "      <td>MDM2</td>\n",
       "      <td>E3 ubiquitin-protein ligase Mdm2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Q02763</th>\n",
       "      <td>TIE2</td>\n",
       "      <td>Angiopoietin-1 receptor</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Q04759</th>\n",
       "      <td>PRKCQ</td>\n",
       "      <td>Protein kinase C theta type</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Q05769</th>\n",
       "      <td>COX2</td>\n",
       "      <td>Prostaglandin G/H synthase 2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Q07343</th>\n",
       "      <td>PDE4B</td>\n",
       "      <td>cAMP-specific 3',5'-cyclic phosphodiesterase 4B</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Q07869</th>\n",
       "      <td>PPARa</td>\n",
       "      <td>Peroxisome proliferator-activated receptor alpha</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Q08881</th>\n",
       "      <td>ITK</td>\n",
       "      <td>Tyrosine-protein kinase ITK/TSK</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Q13464</th>\n",
       "      <td>ROCK-1</td>\n",
       "      <td>Rho-associated protein kinase 1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Q16539</th>\n",
       "      <td>P38</td>\n",
       "      <td>Mitogen-activated protein kinase 14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Q16769</th>\n",
       "      <td>QPCT</td>\n",
       "      <td>Glutaminyl-peptide cyclotransferase</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Q6DE08</th>\n",
       "      <td>AURKB</td>\n",
       "      <td>Aurora kinase B-A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Q8IXJ6</th>\n",
       "      <td>SIRT2</td>\n",
       "      <td>NAD-dependent protein deacetylase sirtuin-2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Q92731</th>\n",
       "      <td>ER-beta</td>\n",
       "      <td>Estrogen receptor beta</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Q92769</th>\n",
       "      <td>HDAC2</td>\n",
       "      <td>Histone deacetylase 2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Q9BY41</th>\n",
       "      <td>HDAC8</td>\n",
       "      <td>Histone deacetylase 8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Q9BYF1</th>\n",
       "      <td>ACE-2</td>\n",
       "      <td>Angiotensin-converting enzyme 2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Q9P1W9</th>\n",
       "      <td>PIM-2</td>\n",
       "      <td>Serine/threonine-protein kinase pim-2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                  index                                               name\n",
       "accession                                                                 \n",
       "P17948           VEGFR1      Vascular endothelial growth factor receptor 1\n",
       "P19793              RXR                   Retinoic acid receptor RXR-alpha\n",
       "P19971               TP                            Thymidine phosphorylase\n",
       "P24941             CDK2                          Cyclin-dependent kinase 2\n",
       "P27907               NA                                      Neuraminidase\n",
       "P28845     11-beta-HSD1     Corticosteroid 11-beta-dehydrogenase isozyme 1\n",
       "P29274              A2A                             Adenosine receptor A2a\n",
       "P31749             AKT1          RAC-alpha serine/threonine-protein kinase\n",
       "P35968           VEGFR2      Vascular endothelial growth factor receptor 2\n",
       "P37231            PPARg   Peroxisome proliferator-activated receptor gamma\n",
       "P43235             CTSK                                        Cathepsin K\n",
       "P45983             JNK1                 Mitogen-activated protein kinase 8\n",
       "P45984             JNK2                 Mitogen-activated protein kinase 9\n",
       "P48736            PI3Kg  Phosphatidylinositol 4,5-bisphosphate 3-kinase...\n",
       "P49137              MK2              MAP kinase-activated protein kinase 2\n",
       "P49841            GSK3B                    Glycogen synthase kinase-3 beta\n",
       "P52333             JAK3                       Tyrosine-protein kinase JAK3\n",
       "P52732            KIF11                         Kinesin-like protein KIF11\n",
       "P53779             JNK3                Mitogen-activated protein kinase 10\n",
       "P54760            EPHB4                           Ephrin type-B receptor 4\n",
       "P55859              PNP                    Purine nucleoside phosphorylase\n",
       "P62942           FKBP1A         Peptidyl-prolyl cis-trans isomerase FKBP1A\n",
       "P78536           ADAM17  Disintegrin and metalloproteinase domain-conta...\n",
       "P80276             ALR2                                   Aldose reductase\n",
       "P9WGR1             INHA      Enoyl-[acyl-carrier-protein] reductase [NADH]\n",
       "Q00987             MDM2                   E3 ubiquitin-protein ligase Mdm2\n",
       "Q02763             TIE2                            Angiopoietin-1 receptor\n",
       "Q04759            PRKCQ                        Protein kinase C theta type\n",
       "Q05769             COX2                       Prostaglandin G/H synthase 2\n",
       "Q07343            PDE4B    cAMP-specific 3',5'-cyclic phosphodiesterase 4B\n",
       "Q07869            PPARa   Peroxisome proliferator-activated receptor alpha\n",
       "Q08881              ITK                    Tyrosine-protein kinase ITK/TSK\n",
       "Q13464           ROCK-1                    Rho-associated protein kinase 1\n",
       "Q16539              P38                Mitogen-activated protein kinase 14\n",
       "Q16769             QPCT                Glutaminyl-peptide cyclotransferase\n",
       "Q6DE08            AURKB                                  Aurora kinase B-A\n",
       "Q8IXJ6            SIRT2        NAD-dependent protein deacetylase sirtuin-2\n",
       "Q92731          ER-beta                             Estrogen receptor beta\n",
       "Q92769            HDAC2                              Histone deacetylase 2\n",
       "Q9BY41            HDAC8                              Histone deacetylase 8\n",
       "Q9BYF1            ACE-2                    Angiotensin-converting enzyme 2\n",
       "Q9P1W9            PIM-2              Serine/threonine-protein kinase pim-2"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dekois_accessions.sort_index()[39:]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Merging the datasets\n",
    "df_targets = pd.concat([dud_accessions, dekois_accessions], axis = 0).sort_index()\n",
    "df_targets = df_targets.iloc[~df_targets.index.duplicated(keep='last')]\n",
    "# Naming the table\n",
    "df_targets.columns = ['name', 'long_name']\n",
    "\n",
    "# Fill a new columns indicating if the target is in a given datset\n",
    "# for dud\n",
    "df_targets.loc[:, 'DUD'] = [1 if i in dud_accessions.index else 0 for i in df_targets.index]\n",
    "# for dekois\n",
    "df_targets.loc[:, 'DEKOIS2'] = [1 if i in dekois_accessions.index else 0 for i in df_targets.index]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>name</th>\n",
       "      <th>long_name</th>\n",
       "      <th>DUD</th>\n",
       "      <th>DEKOIS2</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>accession</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>O14965</th>\n",
       "      <td>AURKA</td>\n",
       "      <td>Aurora kinase A</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>O15530</th>\n",
       "      <td>PDK1</td>\n",
       "      <td>3-phosphoinositide-dependent protein kinase 1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>O76074</th>\n",
       "      <td>PDE5</td>\n",
       "      <td>cGMP-specific 3',5'-cyclic phosphodiesterase</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P00374</th>\n",
       "      <td>DHFR</td>\n",
       "      <td>Dihydrofolate reductase</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P00533</th>\n",
       "      <td>EGFR</td>\n",
       "      <td>Epidermal growth factor receptor</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Q92731</th>\n",
       "      <td>ER-beta</td>\n",
       "      <td>Estrogen receptor beta</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Q92769</th>\n",
       "      <td>HDAC2</td>\n",
       "      <td>Histone deacetylase 2</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Q9BY41</th>\n",
       "      <td>HDAC8</td>\n",
       "      <td>Histone deacetylase 8</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Q9BYF1</th>\n",
       "      <td>ACE-2</td>\n",
       "      <td>Angiotensin-converting enzyme 2</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Q9P1W9</th>\n",
       "      <td>PIM-2</td>\n",
       "      <td>Serine/threonine-protein kinase pim-2</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>91 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "              name                                      long_name  DUD  \\\n",
       "accession                                                                \n",
       "O14965       AURKA                                Aurora kinase A    0   \n",
       "O15530        PDK1  3-phosphoinositide-dependent protein kinase 1    0   \n",
       "O76074        PDE5   cGMP-specific 3',5'-cyclic phosphodiesterase    0   \n",
       "P00374        DHFR                        Dihydrofolate reductase    1   \n",
       "P00533        EGFR               Epidermal growth factor receptor    1   \n",
       "...            ...                                            ...  ...   \n",
       "Q92731     ER-beta                         Estrogen receptor beta    0   \n",
       "Q92769       HDAC2                          Histone deacetylase 2    0   \n",
       "Q9BY41       HDAC8                          Histone deacetylase 8    0   \n",
       "Q9BYF1       ACE-2                Angiotensin-converting enzyme 2    0   \n",
       "Q9P1W9       PIM-2          Serine/threonine-protein kinase pim-2    0   \n",
       "\n",
       "           DEKOIS2  \n",
       "accession           \n",
       "O14965           1  \n",
       "O15530           1  \n",
       "O76074           1  \n",
       "P00374           1  \n",
       "P00533           1  \n",
       "...            ...  \n",
       "Q92731           1  \n",
       "Q92769           1  \n",
       "Q9BY41           1  \n",
       "Q9BYF1           1  \n",
       "Q9P1W9           1  \n",
       "\n",
       "[91 rows x 4 columns]"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_targets"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "20"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(df_targets[df_targets['DUD'] == 1])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {},
   "outputs": [],
   "source": [
    "# df_targets.to_json('targets_table.json')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Get the number of crystals available for each target"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "from bs4 import BeautifulSoup\n",
    "import requests\n",
    "\n",
    "def pdb_ids_from_uniprot(uniprot_id, time_sleep = 2):\n",
    "    r = requests.get('https://www.uniprot.org/uniprot/' + uniprot_id)\n",
    "    soup = BeautifulSoup(r.content)\n",
    "    pdb_tags = soup.find_all('a', {'class': 'pdb'})\n",
    "    \n",
    "    pdb_ids = [tag.get_text() for tag in pdb_tags]\n",
    "    \n",
    "    result_dict = {'pdb_ids': pdb_ids, 'n_crystals': len(pdb_ids)}\n",
    "    \n",
    "    print(uniprot_id, len(pdb_ids))\n",
    "    time.sleep(time_sleep)\n",
    "    return(result_dict)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "O14965 156\n",
      "O15530 68\n",
      "O76074 38\n",
      "P00374 79\n",
      "P00533 213\n",
      "P00734 392\n",
      "P00742 146\n",
      "P00749 146\n",
      "P00750 11\n",
      "P00811 106\n",
      "P00813 2\n",
      "P03366 337\n",
      "P03372 296\n",
      "P04035 22\n",
      "P04058 118\n",
      "P04062 32\n",
      "P04150 43\n",
      "P04183 4\n",
      "P04585 214\n",
      "P04587 71\n",
      "P04626 34\n",
      "P04818 60\n",
      "P05979 30\n",
      "P06239 56\n",
      "P06401 20\n",
      "P06737 18\n",
      "P07550 36\n",
      "P07711 38\n",
      "P07900 300\n",
      "P08069 30\n",
      "P08253 11\n",
      "P09874 67\n",
      "P0C6X7 101\n",
      "P10275 82\n",
      "P10415 27\n",
      "P11217 1\n",
      "P11309 165\n",
      "P11362 66\n",
      "P11509 11\n",
      "P12821 47\n",
      "P12931 64\n",
      "P14061 27\n",
      "P15056 81\n",
      "P15121 145\n",
      "P17948 12\n",
      "P19793 89\n",
      "P19971 4\n",
      "P21964 12\n",
      "P22102 29\n",
      "P22303 51\n",
      "P23219 1\n",
      "P24941 416\n",
      "P27907 5\n",
      "P28845 40\n",
      "P29274 52\n",
      "P31749 32\n",
      "P35354 8\n",
      "P35968 54\n",
      "P37231 225\n",
      "P43235 61\n",
      "P45983 32\n",
      "P45984 2\n",
      "P48736 95\n",
      "P49137 26\n",
      "P49841 87\n",
      "P52333 36\n",
      "P52732 58\n",
      "P53779 51\n",
      "P54760 23\n",
      "P55859 21\n",
      "P62942 51\n",
      "P78536 23\n",
      "P80276 5\n",
      "P9WGR1 97\n",
      "Q00987 111\n",
      "Q02763 15\n",
      "Q04759 8\n",
      "Q05769 51\n",
      "Q07343 43\n",
      "Q07869 21\n",
      "Q08881 36\n",
      "Q13464 27\n",
      "Q16539 242\n",
      "Q16769 26\n",
      "Q6DE08 12\n",
      "Q8IXJ6 31\n",
      "Q92731 32\n",
      "Q92769 8\n",
      "Q9BY41 50\n",
      "Q9BYF1 22\n",
      "Q9P1W9 2\n"
     ]
    }
   ],
   "source": [
    "targets_pdb_ids = {}\n",
    "for uniprot_id in df_targets.index:\n",
    "    targets_pdb_ids[uniprot_id] = pdb_ids_from_uniprot(uniprot_id)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "pd.DataFrame(targets_pdb_ids).T.to_json('targets_PDB_IDS.json')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Old functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 417,
   "metadata": {},
   "outputs": [],
   "source": [
    "# import requests\n",
    "# from Bio import SeqIO\n",
    "# import pickle\n",
    "# import pypdb\n",
    "\n",
    "def _blast_pdb(uniprot_id, identidad = 95):\n",
    "    \n",
    "    file_blast_record = 'target_fastas/' + uniprot_id + '_blast_record.pkl'\n",
    "    if os.path.isfile(file_blast_record):\n",
    "        blast_record = pickle.load(open(file_blast_record, 'rb'))\n",
    "    else:\n",
    "        # Descargamos el fasta de UniProt\n",
    "        url_fasta = requests.get(\"https://www.uniprot.org/uniprot/\" + uniprot_id + \".fasta\")\n",
    "        file_name_fasta = 'target_fastas/' + uniprot_id + '.fasta'\n",
    "        open(file_name_fasta, 'wb').write(url_fasta.content)\n",
    "        # Leemos la secuenciade aminoácidos\n",
    "        fasta_prot = SeqIO.read(open(file_name_fasta),'fasta')\n",
    "        seq_prot = str(fasta_prot.seq)\n",
    "        try:\n",
    "            blast_record = blastPDB(seq_prot)\n",
    "            pickle.dump(blast_record, open(file_blast_record, 'wb'))\n",
    "        except: pass\n",
    "    \n",
    "    pdbids = blast_record.getHits(percent_identity = identidad) # Devuelve un diccionario con cada proteína hit\n",
    "    len(pdbids) # num de proteínas que tienen un 95% de identidad con la secuencia\n",
    "    pdbids_list = list(pdbids.keys()) # lista de pdb_ids de las estructuras de la proteina\n",
    "    print(F'{len(pdbids_list)} hits con una identidad del {identidad}%')\n",
    "    time.sleep(2)\n",
    "    return(pdbids_list)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "metadata": {},
   "outputs": [],
   "source": [
    "# import requests\n",
    "# from Bio import SeqIO\n",
    "# import pickle\n",
    "# import pypdb\n",
    "\n",
    "def _blast_pdb(uniprot_id, identidad = 95):\n",
    "    \n",
    "    file_blast_record = 'target_fastas/' + uniprot_id + '_blast_record.pkl'\n",
    "    if os.path.isfile(file_blast_record):\n",
    "        blast_record = pickle.load(open(file_blast_record, 'rb'))\n",
    "    else:\n",
    "        # Descargamos el fasta de UniProt\n",
    "        url_fasta = requests.get(\"https://www.uniprot.org/uniprot/\" + uniprot_id + \".fasta\")\n",
    "        fasta = ''.join(url_fasta.text.split('\\n')[1:])\n",
    "        seq_prot = fasta\n",
    "        try:\n",
    "            blast_record = blastPDB(seq_prot)\n",
    "            pickle.dump(blast_record, open(file_blast_record, 'wb'))\n",
    "        except: pass\n",
    "    \n",
    "    pdbids = blast_record.getHits(percent_identity = identidad) # Devuelve un diccionario con cada proteína hit\n",
    "    len(pdbids) # num de proteínas que tienen un 95% de identidad con la secuencia\n",
    "    pdbids_list = list(pdbids.keys()) # lista de pdb_ids de las estructuras de la proteina\n",
    "    print(F'{len(pdbids_list)} hits con una identidad del {identidad}%')\n",
    "    time.sleep(2)\n",
    "    return(pdbids_list)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 174,
   "metadata": {},
   "outputs": [],
   "source": [
    "# import pypdb\n",
    "# import untangle\n",
    "# import requests\n",
    "# import pandas as pd\n",
    "# import numpy as np\n",
    "\n",
    "def _blast_pdb(uniprot_id, eCutoff = 0, time_sleep = 2):\n",
    "    url_fasta = requests.get(\"https://www.uniprot.org/uniprot/\" + uniprot_id + \".fasta\")\n",
    "    seq = ''.join(url_fasta.text.split('\\n')[1:])\n",
    "\n",
    "    payload = {\n",
    "                'sequence': seq,\n",
    "                'eCutoff':str(eCutoff),\n",
    "                'matrix':'BLOSUM62',\n",
    "                'outputFormat':'XML' # or 'HTML' or nothing for text format\n",
    "                }\n",
    "\n",
    "    pdbUrl = 'http://www.rcsb.org/pdb/rest/getBlastPDB1'\n",
    "\n",
    "    # Create the url with parameters\n",
    "    r = requests.get(pdbUrl, params = payload)\n",
    "    # Parse \n",
    "    soup = BeautifulSoup(r.content)\n",
    "\n",
    "    pdb_ids = []\n",
    "    for hit in soup.find_all('hit'):\n",
    "        pdb_id = hit.find('hit_def').get_text().split(':')[0]\n",
    "        pdb_ids.append(pdb_id)\n",
    "    #     e_val = float(hit.find('hsp_evalue').get_text())\n",
    "    #     hit_len = int(hit.find('hit_len').get_text())\n",
    "    print(uniprot_id, len(pdb_ids))\n",
    "    \n",
    "    time.sleep(time_sleep)\n",
    "    return({'pdb_ids': pdb_ids, 'n_crystals': len(pdb_ids)})\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "hideCode": false,
    "hideOutput": true,
    "hidePrompt": false
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {
    "hideCode": false,
    "hidePrompt": false
   },
   "outputs": [],
   "source": [
    "dud_targets_names = {\n",
    "    \"ACE\": \"Angiotensin-converting enzyme\",\n",
    "    \"ACHE\": \"Acetylcholine esterase\",\n",
    "    \"ADA\": \"Adenosine deaminase\",\n",
    "    \"ALR2\": \"Aldose reductase\",\n",
    "    \"AmpC\": \"AmpC beta lactamase\",\n",
    "    \"AR\": \"Androgen receptor\",\n",
    "    \"CDK2\": \"Cyclin dependent kinase 2\",\n",
    "    \"COMT\": \"Catechol O-methyltransferase\",\n",
    "    \"COX1\": \"Cyclooxygenase 1\",\n",
    "    \"COX2\": \"Cyclooxygenase 2\",\n",
    "    \"DHFR\": \"Dihydrofolate reductase\",\n",
    "    \"EGFr\": \"Epidermal growth factor receptor kinase\",\n",
    "    \"ER_agonist\": \"Estrogen receptor agonist\",\n",
    "    \"ER_antagonist\": \"Estrogen receptor antagonist\",\n",
    "    \"FGFr1\": \"Fibroblast growth factor receptor kinase\",\n",
    "    \"FXa\": \"Factor Xa\",\n",
    "    \"GART\": \"glycinamide ribonucleotide transformylase\",\n",
    "    \"GPB\": \"Glycogen phosphorylase beta\",\n",
    "    \"GR\": \"Glutocorticoid receptor\",\n",
    "    \"HIVPR\": \"HIV protease\",\n",
    "    \"HIVRT\": \"HIV reverse transcriptase\",\n",
    "    \"HMGR\": \"Hydroxymethylglutaryl-CoA reductase\",\n",
    "    \"HSP90\": \"Human heat shock protein 90 kinase\",\n",
    "    \"InhA\": \"Enoyl ACP reductase\",\n",
    "    \"MR\": \"Mineralcorticoid receptor\",\n",
    "    \"NA\": \"Neuraminidase\",\n",
    "    \"P38\": \"P38 mitogen activated protein kinase\",\n",
    "    \"PARP\": \"Poly(ADP-ribose) polymerase\",\n",
    "    \"PDE5\": \"Phosphodiesterase V\",\n",
    "    \"PDGFrb\": \"Platlet derived growth factor receptor kinase\",\n",
    "    \"PNP\": \"Purine nucleoside phosphorylase\",\n",
    "    \"PPARg\": \"Peroxisome proliferator activated receptor gamma\",\n",
    "    \"PR\": \"Progesterone receptor\",\n",
    "    \"RXRa\": \"Retinoic X receptor alpha\",\n",
    "    \"SAHH\": \"S-adenosyl-homocysteine hydrolase\",\n",
    "    \"SRC\": \"Tyrosine kinase SRC\",\n",
    "    \"Thrombin\": \"Thrombin\",\n",
    "    \"TK\": \"Thymidine kinase\",\n",
    "    \"Trypsin\": \"Trypsin\",\n",
    "    \"VEGFr2\": \"Vascular endothelial growth factor receptor kinase\"\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "dekois_targets_names = {\n",
    "    \"11-beta-HSD1\": \"3tfq\",\n",
    "    \"17-beta-HSD1\": \"3klm\",\n",
    "    \"A2A\": \"3eml\",\n",
    "    \"ACE\": \"1uze\",\n",
    "    \"ACE-2\": \"1r4l\",\n",
    "    \"ACHE\": \"1eve\",\n",
    "    \"ADAM17\": \"3ewj\",\n",
    "    \"ADRB2\": \"3ny9\",\n",
    "    \"AKT1\": \"3qkl\",\n",
    "    \"ALR2\": \"1ah3\",\n",
    "    \"AR\": \"1e3g\",\n",
    "    \"AURKA\": \"3fdn\",\n",
    "    \"AURKB\": \"2vgo\",\n",
    "    \"Bcl-2\": \"2w3l\",\n",
    "    \"BRAF\": \"3skc\",\n",
    "    \"CATL\": \"3bc3\",\n",
    "    \"CDK2\": \"1ckp\",\n",
    "    \"COX1\": \"3kk6\",\n",
    "    \"COX2\": \"1cx2\",\n",
    "    \"CTSK\": \"3kx1\",\n",
    "    \"CYP2A6\": \"1z11\",\n",
    "    \"DHFR\": \"1s3v\",\n",
    "    \"EGFR\": \"1m17\",\n",
    "    \"EPHB4\": \"2vwz\",\n",
    "    \"ERBB2\": \"3pp0\",\n",
    "    \"ER-beta\": \"3oll\",\n",
    "    \"FGFR1\": \"1agw\",\n",
    "    \"FKBP1A\": \"2dg3\",\n",
    "    \"FXA\": \"1f0r\",\n",
    "    \"GBA\": \"2wcg\",\n",
    "    \"GR\": \"1nhz\",\n",
    "    \"GSK3B\": \"3i4b\",\n",
    "    \"HDAC2\": \"3max\",\n",
    "    \"HDAC8\": \"3sff\",\n",
    "    \"HIV1PR\": \"3nu3\",\n",
    "    \"HIV1RT\": \"1s6p\",\n",
    "    \"HMGA\": \"1hw8\",\n",
    "    \"HSP90\": \"1uy6\",\n",
    "    \"IGF1R\": \"3nw7\",\n",
    "    \"INHA\": \"1p44\",\n",
    "    \"ITK\": \"3mj1\",\n",
    "    \"JAK3\": \"3lxl\",\n",
    "    \"JNK1\": \"3elj\",\n",
    "    \"JNK2\": \"3npc\",\n",
    "    \"JNK3\": \"2b1p\",\n",
    "    \"KIF11\": \"3k5e\",\n",
    "    \"Lck\": \"3mpm\",\n",
    "    \"MDM2\": \"3lbk\",\n",
    "    \"MK2\": \"3kc3\",\n",
    "    \"MMP2\": \"1hov\",\n",
    "    \"NA\": \"1a4g\",\n",
    "    \"P38\": \"1ouk\",\n",
    "    \"PARP-1\": \"3l3m\",\n",
    "    \"PDE4B\": \"3frg\",\n",
    "    \"PDE5\": \"1xp0\",\n",
    "    \"PDK1\": \"2xch\",\n",
    "    \"PI3Kg\": \"3dbs\",\n",
    "    \"PIM-1\": \"3r04\",\n",
    "    \"PIM-2\": \"2iwi\",\n",
    "    \"PNP\": \"1b8o\",\n",
    "    \"PPARa\": \"2p54\",\n",
    "    \"PPARg\": \"1fm9\",\n",
    "    \"PR\": \"2w8y\",\n",
    "    \"PRKCQ\": \"1xjd\",\n",
    "    \"PYGL-in\": \"1xoi\",\n",
    "    \"PYGL-out\": \"3dds\",\n",
    "    \"QPCT\": \"2afx\",\n",
    "    \"ROCK-1\": \"3v8s\",\n",
    "    \"RXR\": \"2p1t\",\n",
    "    \"SARS-HCoV\": \"2z94\",\n",
    "    \"SIRT2\": \"1j8f\",\n",
    "    \"SRC\": \"2src\",\n",
    "    \"Thrombin\": \"3rm2\",\n",
    "    \"TIE2\": \"2oo8\",\n",
    "    \"TK\": \"1w4r\",\n",
    "    \"TP\": \"1uou\",\n",
    "    \"TPA\": \"1a5h\",\n",
    "    \"TS\": \"1i00\",\n",
    "    \"uPA\": \"3mhw\",\n",
    "    \"VEGFR1\": \"3hng\",\n",
    "    \"VEGFR2\": \"3c7q\"\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "81"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(dekois_targets_names)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "hide_code_all_hidden": false,
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.6"
  },
  "toc": {
   "base_numbering": 1,
   "nav_menu": {},
   "number_sections": true,
   "sideBar": true,
   "skip_h1_title": false,
   "title_cell": "Table of Contents",
   "title_sidebar": "Contents",
   "toc_cell": false,
   "toc_position": {},
   "toc_section_display": true,
   "toc_window_display": false
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
